{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-light\"><h2 style=\"font-size:36px\"style=\"padding:1px\"><center>Take Home - Notebook Developer Submission </center></h2><center> All answers contain the reasoning behind the choices and route/reasoning taken. </div>  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><h2 style=\"font-size:16px\"style=\"padding:1px\"><center>Q1. READING BEST-PRACTICE PUBLICATIONS </center></h2>\n",
    "<p>When developing a new pipeline, or improve an existing one, a good starting point is a review paper that discusses the best practices. If you need to develop a pipeline, or improve an existing one, identify a single publication that would like to follow. You don’t have to read the full paper! Just search for the relevant papers, take a brief look, and select the one you would want to follow in detail. Share the link to the publication. During your search, you may have also found other good papers. Why did you pick the first one? Which papers were number 2 or 3? For this exercise, you may pick either RNA-Seq or ChIP-Seq or single nucleotide variant calling.\n",
    "Using Python, Pandas, Numpy and other common dependencies. All answers contain the reasoning behind the choices and route taken. </div> \n",
    "\n",
    "<div class=\"alert alert-block alert-warning\"><h2 style=\"font-size:16px\"style=\"padding:1px\"><center>Q1. ANSWER: </center></h2>\n",
    "<h4>RNA-SEQ</h4>\n",
    "<h6>1</h6> A survey of best practices for RNA-seq data analysis. Conesa A, Madrigal P, Tarazona S, Gomez-Cabrero D, Cervera A, McPherson A, Szcześniak MW, Gaffney DJ, Elo LL, Zhang X, Mortazavi A. Genome Biol. 2016 Jan 26;17:13. doi: 10.1186/s13059-016-0881-8. Erratum in: Genome Biol. 2016;17(1):181. PMID: 26813401; PMCID: PMC4728800.\n",
    "Direct Link: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4728800/\n",
    "\n",
    "The paper breaks the process (RNA-Seq) in modules. These modules are then studied for it's characteristics for the signs that point either to a good pipeline or the reasons along the way for the need to improvement of a ‘faulty’ pipeline.\n",
    "\n",
    "<h6>2</h6>\n",
    "Improving the Flexibility of RNA-Seq Data Analysis Pipelines.Phan JH, Wu PY, Wang MD.  IEEE Int Workshop Genomic Signal Process Stat. 2012 Dec;2012:70-73. doi: 10.1109/GENSIPS.2012.6507729. PMID: 27536420; PMCID: PMC4985025.\n",
    "Direct Link: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4985025/\n",
    "\n",
    "<h6>3</h6>\n",
    "Computational Testing for Automated Preprocessing 2: Practical Demonstration of a System for Scientific Data-Processing Workflow Management for High-Volume EEG. Cowley BU, Korpela J. Front Neurosci. 2018 Apr 10;12:236. doi: 10.3389/fnins.2018.00236. PMID: 29692705; PMCID: PMC5902528.\n",
    "Direct Link:https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5902528/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><h2 style=\"font-size:16px\"style=\"padding:1px\"><center>Q2. EVALUATING TOOLS </center></h2>\n",
    "<p> A common step in many pipelines is pre-processing reads before alignment, to remove low quality bases and trim adapters. There are multiple tools available for this purpose, e.g., TrimGalore, fastp and Atropos.\n",
    "I.\tWhich among the 3 would you use? (or would you use a different tool?)\n",
    "II.\tWhat are the pros and cons of each?</div> \n",
    "\n",
    "<div class=\"alert alert-block alert-warning\"><h2 style=\"font-size:16px\"style=\"padding:1px\"><center>Q2. ANSWER </center></h2>\n",
    "<p> I - FASTP\n",
    "<p> <br>\n",
    "<h4>II</h4>\n",
    "    \n",
    "<h4>TrimGalore</h4> <br>\n",
    "Cons:<br>\n",
    "Requires separate installation of Cutadapt and FastQC.\n",
    "Adapter auto-detection even if no adaptor sequence is supplied.Yet the adaptor auto-detection can be overwritten and manually supplied.\n",
    "<p>\n",
    "Pros: <br>\n",
    "One process in two-steps, it's able to do this by wrapping around two separate tools (Cutadapt and FastQC) joining them in one single process inside TrimGalore which takes output from the first to the second without a need for extra coding steps. \n",
    "<p>\n",
    "<h4>FASTP<br></h4>\n",
    "Multi-threading<br>\n",
    "2 to 5 faster times than FASTQ<br>\n",
    "It includes most features of FASTQC + Cutadapt + Trimmomatic + AfterQC while running 2–5 times faster than any of them alone.<br>\n",
    "<p>\n",
    "<h4>Atropos<br></h4>\n",
    "Artificially methylated bases at the 3′ ends of reads.<br>\n",
    "paired-end read trimming accuracy that is superior to other state-of-the-art tools, and it is among the fastest read trimming tools when a moderate number of parallel execution threads are used. <br>\n",
    "Achieves a performance increase that is roughly linear with the number of threads used, making it the fastest tool when eight or more threads are available.<br>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><h2 style=\"font-size:16px\"style=\"padding:1px\"><center>Q3. VALIDATING PIPELINES </center></h2>\n",
    "<p>When new users analyze their data on Basepair, a question they have is: “can I trust these results?” How will you validate a pipeline to show the accuracy? You don’t have to validate the pipeline, just provide an outline of how you will go about it. For this exercise, you may pick either RNA-Seq or ChIP-Seq or single nucleotide variant calling.\n",
    "</div> \n",
    "\n",
    "<div class=\"alert alert-block alert-warning\"><h2 style=\"font-size:16px\"style=\"padding:0px\"> ANSWER </h2>\n",
    "Quality checks start with the pre-processing aspects of the analysis. Certain criteria have to be met for a significant result.\n",
    "Experiment validation and experiment quality checks: Trimming, poly A, GC ratio, percentage of read alignment. Once these are within known standards the sample is accepted as ‘validated’ and not the probable source of any issues.\n",
    "\n",
    "Depending on the desired analysis certain information, data, metadata and tools need to be available. For example:\n",
    "<p>  ● Enrichment Analysis: Background databases need to be available and the User analyzed data needs to follow the same standards.\n",
    "<p>  ● Principal Components Analysis:The number of components areas in which over a total of 80% of features will fall need to be at most four components for most visualization techniques. If a 3D view is available up to 6 components can be acessed.\n",
    "Usually a 2D view with up to four components is used. If those four main components do not account for over 80% of the components either reduced total % found on main components areas, expanding to 3D and obtaining another 2 dimensions of components, further data cleaning or the total % found into the main components can be used as a clear indication of issues that need to be resolved in the analysis.\n",
    "</div> \n",
    "\n",
    "<div class=\"alert alert-block alert-warning\"><h2 style=\"font-size:16px\"style=\"padding:1px\">Program/Script Pipeline Validation and Quality Checks:</h2>\n",
    "<p>\n",
    "\n",
    "<p> <h3> 1. UNIT TESTING (DATA SIDE VALIDATION)</h3> <p>\n",
    "Unit testing allows the pipeline to be broken down into modules and have each of its modules tested for the acceptable input, values and necessary outputs. <p>Some of these include testing: Data type, Data size,Data structure and values.  \n",
    "<p> Singleton Issue Alerts: These are Scripts that exist parallel to the main pipeline and called by the pipeline to perform a check, if validated continue pipeline. If not-valid or pre-defined singleton issue definitions are not met: <p>\n",
    "        <h5>STOP ALERT  </h5><p>\n",
    "            Contains: <p> ISSUE-NAME and ISSUE-TYPE-SPECIFIC-ID; <p> Possible reasons why ISSUE-NAME usually occurs; <p> Possible conducts to be taken.<p>\n",
    "            Is documented:<p>\n",
    "            Keeps issue-name, issue-specific-ID, date and time, <p>\n",
    "            and Sample(s) ID: \n",
    "                <p> * Assigned automatically per sample in run. Each pipeline run is one analysis and the samples contained in that batch are also numbered and identified. \n",
    "                <p> * A pipeline can be used with 1 to unlimited number of samples (unless maximum samples per batch have been limited by  basepair.)\n",
    "                 <p>   * Assigned as run-ID-sample-ID which can be used to see all the run ID pre-sets, type of analysis and other information, the sample-ID can be used to gather all sample specific information. <p>\n",
    "                    \n",
    "<p> <h5> REASONING FOR STOP ALERT:  </h5><p> Detailing why the basepair platform stops when this issue occurs. <p> \n",
    "Provides link to literature on issue. Provides link to possible solutions.<p>\n",
    " * Some of these solutions can be made available with a new button that appears in the reasoning if a script able to correct the issue is available.\n",
    "<p>\n",
    "<h6> OR IGNORE ALERT and CONTINUE ANALYSIS BUTTON. </h6>\n",
    "Which gives the user overwrite control. <p>\n",
    "<p>  <p>\n",
    "Most of these are performed by tools but experiment-user specific code checks might be added to an automated analysis. Singleton issue validation of: Percentage of ____ GC content. Presence of Adaptor. Fragment Size. ETC\n",
    "<p>  <p>\n",
    "    <h3>     2. GOLD STANDARD </h3>\n",
    "                Literature available data that is recognized as a gold standard for assay type that have known results for \n",
    "                known pipelines. The gold standard data will run throught the new proposed pipeline which needs to provide \n",
    "                similar or better  results than those pipelines tested in the literature.\n",
    "<p>\n",
    "<p>  \n",
    "    <h3>      3. VISUALIZATION </h3>\n",
    "                Graphs and Figures can be used to look for issues in the data/analysis/ predetermined variables and/or on the \n",
    "                programing scripts.\n",
    "<p>\n",
    "<p>  \n",
    "     <h3>     4. UNIT TEST (PROGRAMING SIDE VALIDATION) </h3>\n",
    "                Break the pipeline into single task modules and test each of them. \n",
    "                For example:Import data module, Clean data-standardize module, Create Matrix module,Calculate distance module, VisualizationT1 module, VisualizationT2 module. </div> \n",
    "<div class=\"alert alert-block alert-warning\"><h2 style=\"font-size:10px\"style=\"padding:0px\"></h2>\n",
    "When re-assembling the pipeline the end of each module can contain de-tour. Meaning the module output has correct data, type and is working as validated. The de-tour is the parallel checks and analysis specific-special tests and/or commands. For example a de-tour can higher restrictive criteria before using that module output to start a new sideways route and that same output will also be used to continue the main pipeline. \n",
    "\n",
    "In this one Only one main pipeline exists and from that always same input location/script/computation the other possibilities branch out. This saves in computation the data is cleaned, wrangled, organized and standardized just one time for any subsequential analysis. \n",
    "\n",
    "Once the sample data in uploaded the platform does not need to ever perform this step for the same sample unless the user deletes the analysis read data file.\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\"><h2 style=\"font-size:16px\"style=\"padding:1px\">User Stories</h2>\n",
    "<p>\n",
    "\n",
    "USER STORIES<br>\n",
    "Another important part of developing and validating a pipeline is the use of user stories. Inside academia user stories are almost never given too much thought, this is true especially for higher level and appliance science.\n",
    "\n",
    "A user story has some general basic components:\n",
    "\n",
    "User knowledge: \n",
    "A PhD in the field will need a different view of the information when compared to\n",
    "for example a governmental user that has been tasked with evaluating if the invested governmental resources\n",
    "were able to generate a real outcome. In this example case the PhD can browse quickly over the menus, settings and options.\n",
    "While a government official might turn desperate without a general roadmap of the project with simple explanations and clear division of data, research progress and its future.\n",
    "\n",
    "User wants: \n",
    "What do the users want from the tools/website/platform.\n",
    "How did they end up at that location asking a question?\n",
    "\n",
    "User issues: \n",
    "What are the issues encountered; For instance, couldn't understand settings for (_____), can a specific request for a change to code be submitted, among so many others.\n",
    "\n",
    "A Necessary Epic\n",
    "An epic is necessary for the development team. What we call an epic can be defined as a large collection of users stories.\n",
    "This epic can be seen as multiple facets of the same 'point' by respecting the view, perception and even distortion\n",
    "that is a part of any story a development team can deliver a truly outstanding product.\n",
    "That will be able to supply the required information from a fifth grade assignment, to a news piece and going all the way up to the highest level of science which needs discovery, ideas, possibilities that needs to take assumptions and comparable insights to see and write, to document and develop what does not exist yet.\n",
    "This maybe a new branch of science, a revolutionary drug or a simple point change to something common that creates a new view, a new informed understanding, based and motivated by scientific guidelines and thus stepping into the future.\n",
    "\n",
    "While it may sound glamorous and high-profile even, user stories are generally just a few simple phrases long. It’s up to the development and research team to use each of those stories to branch out, create resulting a final built capable of information a large range of knowledge, capable of providing real & significant biological/scientific meaning, clearly justify assumptions and readily present the background and bases for all steps, actions and decisions that from the first click all the way to final results.  \n",
    "\n",
    "Just as a matter of curiosity, the following are some examples of user stories:  <br>\n",
    "- As a user, I would like to be able to quickly write down my thoughts so that I can return to this specific location/information and analysis step. <br>\n",
    "- As a user, I would like to be able to compare across final results in clean and easy to see visualization. To view and compare multiple tries, backgrounds and be able to slowly build a large 'dashboard' around my work. <br>\n",
    "- As a user, I would like to have a general tool that counts my most common issues/problems/alerts in a report, showing sample data and metadata. A problem/issue report that would look what is common among those. <br>\n",
    "- As a user, I would like to have an analysis pipeline diagram automatically generated. That can be specific to my data, work, question and other needs. ** This tell the development team to create a hierarchical tree of the analysis pipeline, create an overview of the user data and its metadata categories. The initial diagram will be complete, with each pipeline step, the changes between one step and the next. The other possible routes that the user-specific data and analysis at the specific-point can take, among others. The hierarchical view will allowed the user to select/unselect steps, metadata, pipeline branching which can be taken-out or reduced. Among many other possible features. <br>\n",
    "<br>One interesting feature I would add: Show point were this analysis vary its values, definitions, limits, etc and create a table displaying that step possible variations informing significant, noise, reasoning while keeping in the analysis documentation a registered pipeline decision that was based on science, literature, known analysis tools/limits, this table can dispel any reasons to doubt the researcher choices and validates the result.\n",
    "<p>\n",
    "- As a user, I would like to study the analysis branches and its recommendations to design all my project as cleanly as possible, starting at sample and assay protocols. Which are also when available part of the platform information.\n",
    "- As a user I would like to have a real information-science heavy result and final report. \n",
    "<p>* This can and will be available by choosing next to the steps and metadata to add or not literature and background information. The selected fields and data will use known governmental repositories and databases to create the Almighty-report. Default settings will search for the last 10 years, group, clean, standardize and use Natural Language Processing to created correlations.\n",
    "<p>\n",
    "Which means all the data and its result has proven, tested, validated, corroborated and multi-feature-variable testing available. With The background grouped by its assay, tissue, cell line, disease, drug or other user defined words or three-word phrase. These will also be available in network view for each category and as a full network. From this network a highlight red ! marks the places in that due to network, its nodes, distances and processes have been found to have higher significance (as a theory and a new possibility.)\n",
    "<p>\n",
    "*** This can be easily seen due to modules, their Lego-assembly capabilities, constrains, backgrounds, limits, natural language processing and so on as the beginning, the structure ready-like beginning of machine learning taking the governmental repositories, literature and possibilities to find, adapt, suggest and create (since it will be able to study the successful creative science processes.)\n",
    "<p>\n",
    "- As a user, I would like to be able to collaborate with other users that have similar data, or write down what is the data, assay, type, cells, disease, etc.<br> I am open to collaborate. * Any data, diagram-schema, user-specific-code-changes that are part of a user who wishes to collaborate work will be round-searched once every 15 days, all matches returned and an automatic email sent from those open to collaborate to others who are also collaboration-open.<p>\n",
    "<p>\n",
    "If a User goes to the collaboration page select the project, add relevant information, any restrictions (for instance country due to source of funding), among others will connect those willing to collaborate with no maximum collaborators/institutions numbers. If any sample is  found to be repeated (all data and metadata is cross-referenced) the repeated sample will be evaluated and a final value found by the best protocols and validation methods found, this will assign the general sample collaboration characteristics and give the specific sample result a slightly higher significance value since it’s been tested in different locations, labs, reagents and researchers. This sample needs to be given a higher weight when compared to a single sample with that was only tested once due to for example not-enough sample recovered. This second case cannot be taken with equal weights to thoroughly validated, triplicate tested in multi-locations.<p>\n",
    "<p>\n",
    "* All this information of weights, reasoning and the step-by-step calculations can be easily viewed, sorted, used to define sample-types that being too low-weighted and badly represented in the full sample-set to create an open call to supplement these issues. A sample of weaker weight might be the most important, informative, significant find but that can only be tested if a way to add significance to that specific-sample is found. Some bioinformatics tools can be of help with this.\n",
    "And many more.\n",
    "<p>\n",
    "<p>\n",
    "Continued after the last question answer.\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><h2 style=\"font-size:16px\"style=\"padding:1px\"><center>Q4. PROGRAMMING </center></h2>\n",
    "In a building, people have either cats or dogs. The animals are either white or black.\n",
    "1.\tGiven a list of animals and another list of colors, where the animal and colors are in matched order, count the number of animal of a given color.\n",
    "●\tPreferably complete this assignment in Python and send the source code.\n",
    "</div> \n",
    "\n",
    "<div class=\"alert alert-block alert-warning\"><h2 style=\"font-size:20px\"style=\"padding:1px\"><center>Q4. ANSWER </center></h2>\n",
    "<p>\n",
    "Lists are a type of (linear) data type in which order is established and can only be changed by resetting the list. That being true the order in which a list is presented can matter for the subsequent analysis. The way to count the animals also depends on the desired outcomes and it's uses. \n",
    "I would define both sets of information (animal & color) as two lists. Since lists have a defined order this would make it easy to  assign animal  to color. The first value of lsta would be assigned to the first value of lstb and so on. This is achieved easily by staking the two list and thus creating a dataframe.\n",
    "<br><br>\n",
    "Since the data/information  is taken as a:<br>\n",
    "1x Number_of_animals <br>\n",
    "1x Number_of_colors.<br>\n",
    "This those lists can be stacked creating a 2 x Number_of_animals dataframe with unique index fr future analysis steps.<br>\n",
    "<p><p>\n",
    "-- As a validation step, a simple but necessary unit test can be performed that will validade and increase the significance and trustworthiness of the result. All programing steps should be validated via unit tests and this is a perfect point for it. \n",
    "<p>\n",
    "          This unit test will test the programing, the data, metadata collected and informed) both lists can be have it's number of values contained in the set-list compared.  This test is easy to perform by supplying the programming script with the two lists that will have it's N x M dimensions compared, and those N x M dimensions should be equal.<br>\n",
    "<p>\n",
    "        If the number of animals in lsta is equal the number of colors in lstb than the analysis can continue otherwise an error message should be display saying the number of values of both lists supplied are different and therefore some data\\ manipulation/wrangling is necessary. <br>\n",
    "<p>\n",
    "    It is suggested that for missing colors a new color category is created as NA and for all animals that do not contain the color information the NA category is added as color. This new category will also return the total number of animals without the colorvalue assigned. <br>\n",
    "<p><p>\n",
    "    The NA category is only informative and cannot be taken as a single color thus it should not be displayed in the main analysis result when comparing colors and animals and should be displayed separately since it cannot be displayed as a direct relationship to single animal-single color rather is the unknown animal and/or color and should be highlighted in a report table showing all NA with the information available and the position it occupied in the original list.Informing the users that the information in this particular table was not  included in the analysis.  The metadata is the information, description and all other known characteristics of the measured data that can be used to enrich the data.<br>\n",
    "For example knowing that black cats in the specific region are mostly males and weight over 5kg can make this result more significant. This might be due to the genetics of the black cats in that region have higher survival rates for males, thus more males reach adulthood. Which is important for any further studies delimitation, this would raise questions such as: Is there any Y related advantages that are better suited for the location, if there are mutation on a Y gene that causes the carriers to grow bigger and heavier. \n",
    "<p>\n",
    "When compared to the same species (if available) somewhere does this difference becomes significant.<br>    \n",
    "It can also be used for enrichment analysis, for example when comparing color, weight, and etc between the population being studied and other similar populations can be ‘inferred’ as a possible step. \n",
    "<p><p>     \n",
    "If the most significant characteristics found when analysing the data and metadata, usually the ones that fall into the 3 to 4 main components (in a main components analysis) is compared to other known black cats around the globe by region the characteristics of other populations that are similar to the population under study can be taken in as assumption for further study. For example: If an island in Japan have similar occurrence of male black cats with slight deviation up on their weight, what else is known about the Japanese island population that can be taken as possible to the current study. Thus suggesting the next direction of the research with clear data, enrichment analysis, higher significance and real biological meaning. All easily seem by principal components analysis, tables, pizza charts and direct comparisons and their calculated significance. \n",
    "<p>           \n",
    "All of this guides, motivates and justifies the research, next steps, deals with noise in the data by establishing a threshold of similar data and measurements that is validated on the literature. \n",
    "<p>          \n",
    "All of this steps, data, collection steps and data wrangling can and does justify the project and the need for further tests and analysis. It also suggests reasons for concern, possible nutrient depletion and even possible medications to be used for example black cats losing too much fur since the known health loss is known and the genes for the cat fur coat are known, a differential gene expression analysis of those genes might show for instance a fur geneset for a much colder region and thus raise the hypothesis of temperature regulation of fur shedding.\n",
    " <p>           \n",
    "If found to be true a fur thinning medication might be suggested as a solution, finner fur resulting in lower body temperature and less shedding. In this case, the data is the specific-animal and the metadata available the color. \n",
    " <p>           \n",
    "Other examples of metadata include for example animal weight, age, species, single or multicolor, male or female, carnivore/herbivore or omnivore, among other caracteristics.\n",
    "<p>            \n",
    "It's important to inform the user that metadata is important and without it a result has little to no significant biological meaning. Thinking of subsequential analysis is always good to find ways to facilitate the handling of data, this way I would use the two lists created from probably two text files and use them to create a pandas dataframe.\n",
    "\n",
    "</div> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 380,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import csv\n",
    "import sys\n",
    "\n",
    "from pandas import *\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 381,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['cat', 'dog', 'cat', 'cat', 'dog', 'cat', 'cat']\n",
      "['white', 'black', 'black', 'white', 'black', 'white', 'white']\n",
      "\n",
      "Number of Animals =  7\n",
      "Number of Colors =  7\n",
      "Equal lenght lists.\n",
      "\n",
      "  Animal  Color\n",
      "0    cat  white\n",
      "1    dog  black\n",
      "2    cat  black\n",
      "3    cat  white\n",
      "4    dog  black\n",
      "5    cat  white\n",
      "6    cat  white\n"
     ]
    }
   ],
   "source": [
    "with open(\"Animals1.txt\") as f1, open(\"Colors1.txt\") as f2:\n",
    "        line_file1 = f1.readline()\n",
    "        line_file2 = f2.readline()\n",
    "        #print(line_file1, line_file2)\n",
    "        \n",
    "        Animalwords = line_file1.split() \n",
    "        Colorswords = line_file2.split() \n",
    "        print(Animalwords) \n",
    "        print(Colorswords)\n",
    "        print(\"\\nNumber of Animals = \", len((Animalwords)))\n",
    "        print(\"Number of Colors = \", len((Colorswords)))\n",
    "        \n",
    "        if len(Animalwords) != len(Colorswords):\n",
    "                print(\"Different size lists. PIPELINE STOP ALERT.\\n\")\n",
    "                print(\"Txt files contain diferente number of values. Please add NA to any empty cells.\\n\")\n",
    "                print(\"The pipeline has not been adapted to insert this TAG to maintain the user control over the supplied data, but the user file can be automatically given the NA tag/value by uploading the original to the USER DATA CONTROL PAGE.\")\n",
    "                print(\"If your data contain empty cells this step can be performed by supplying the file location and hitting the replace empty button. Which will take your file add NA to all empty cells and give you a new output named NA-Original_File_Name.txt. This script will also clean any trailling lines left behind by csv files and tables.\")\n",
    "\n",
    "        if len(Animalwords) == len(Colorswords):\n",
    "            #print(\"Valid\")\n",
    "            print(\"Equal lenght lists.\\n\")\n",
    "            \n",
    "            ## Stack both lists and create a 2N x M python pandas dataframe. Name: labels & columns. Use unique index.\n",
    "            #dataframe = pd.DataFrame(np.column_stack([Animalwords, Colorswords]), columns=[\"Animal\", \"Color\"])\n",
    "            #print(dataframe)\n",
    "            \n",
    "            df = pd.DataFrame(list(zip(Animalwords, Colorswords)), columns =['Animal', 'Color']) \n",
    "            print(df)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 375,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Animal</th>\n",
       "      <th>Color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>cat</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>dog</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>cat</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>cat</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>dog</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>cat</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>cat</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Animal  Color\n",
       "0    cat  white\n",
       "1    dog  black\n",
       "2    cat  black\n",
       "3    cat  white\n",
       "4    dog  black\n",
       "5    cat  white\n",
       "6    cat  white"
      ]
     },
     "execution_count": 375,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2D dataframe visualization.\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 376,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Animal  Color\n",
       "cat     white    4\n",
       "        black    1\n",
       "dog     black    2\n",
       "Name: Color, dtype: int64"
      ]
     },
     "execution_count": 376,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Aggregating by animal and color for final Count\n",
    "df.groupby('Animal')['Color'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 377,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Animal</th>\n",
       "      <th>Color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">cat</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>dog</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [(cat, black), (cat, white), (dog, black)]"
      ]
     },
     "execution_count": 377,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# To visualize different colors per animal.\n",
    "df.groupby(['Animal', 'Color']).count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This code works for any number of animals, colors and list lenght."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><h2 style=\"font-size:16px\"style=\"padding:1px\"><center>Q5. VISUALIZATION </center></h2>\n",
    "<p> We ordered a new wireless router to improve the Internet reliability. While everybody feels that the new router, we want to be data-driven. We ran 20 pings with the old and new routers. The data is shown below. Create a single figure with these data to convince some one if the new router is better than the old one. Note that the target audience is a layperson. Preferably complete this assignment in Python, and send the source code and the figure.\n",
    "</div>\n",
    "<div class=\"alert alert-block alert-warning\"><h2 style=\"font-size:16px\"style=\"padding:1px\"><center>Q5. ANSWER </center></h2>\n",
    "<p> \n",
    "I would take the information that is truly informative given the question. The time of response for each router. Simply take the measured points and use them to plot two lines in a graph. Highlighting the time of response of the new modem compared to the old modem. This seems to be the type of answer this question is looking for. \n",
    "\n",
    "But simple the audience has no idea of any of that means I wouldn’t give that as the final answer for the audience. The main idea is to obviously and clearly show to the audience that the new router is faster and explain why this is important and even better why they should care about that response time and expend money on a new router.\n",
    "\n",
    "To convince someone that the new router is better and if they are completely acknowledge of the subject instead of any graph or visualization I would simply inform them of the average response time of one compared to the other. \n",
    "\n",
    "Get a figure with ROUTER-NEW AVERAGE  next to ROUTER-OLD AVERAGE. This figure would simply face a picture of the router with the average ping response time bellow which I would call ´how fast’ and on the side a picture of the older router and the response time bellow.\n",
    "\n",
    "Under this this figure display I would Have the following information written: YOUR NEW ROUTER WOULD BE _____ % FASTER than your older router. In simple terms this means your internet would ‘work’ that _________ %  amount faster.\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each measurement ping and time for round-trip response are here taken as (x,y) one point will be graphed on the cartesian plane and a line drawn for the old_router.\n",
    "A second line will be plotted in the same manner for the new router."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 350,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>F</th>\n",
       "      <th>G</th>\n",
       "      <th>H</th>\n",
       "      <th>I</th>\n",
       "      <th>Old</th>\n",
       "      <th>L</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>0</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>11.912</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>1</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>21.098</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>2</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>7.113</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>3</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>7.033</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>4</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>5.456</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>5</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>78.941</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>6</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>10.592</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>7</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>7.169</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>8</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>4.850</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>9</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>11.244</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>10</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>4.920</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>11</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>104.141</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>12</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>5.090</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>13</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>5.787</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>14</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>6.098</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>15</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>7.451</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>16</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>9.982</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>17</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>5.462</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>18</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>15.061</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>19</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>6.701</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     A      B     C              D         E   F    G   H     I      Old   L\n",
       "0   64  bytes  from  172.217.10.4:  icmp_seq   0  ttl  51  time   11.912  ms\n",
       "1   64  bytes  from  172.217.10.4:  icmp_seq   1  ttl  51  time   21.098  ms\n",
       "2   64  bytes  from  172.217.10.4:  icmp_seq   2  ttl  51  time    7.113  ms\n",
       "3   64  bytes  from  172.217.10.4:  icmp_seq   3  ttl  51  time    7.033  ms\n",
       "4   64  bytes  from  172.217.10.4:  icmp_seq   4  ttl  51  time    5.456  ms\n",
       "5   64  bytes  from  172.217.10.4:  icmp_seq   5  ttl  51  time   78.941  ms\n",
       "6   64  bytes  from  172.217.10.4:  icmp_seq   6  ttl  51  time   10.592  ms\n",
       "7   64  bytes  from  172.217.10.4:  icmp_seq   7  ttl  51  time    7.169  ms\n",
       "8   64  bytes  from  172.217.10.4:  icmp_seq   8  ttl  51  time    4.850  ms\n",
       "9   64  bytes  from  172.217.10.4:  icmp_seq   9  ttl  51  time   11.244  ms\n",
       "10  64  bytes  from  172.217.10.4:  icmp_seq  10  ttl  51  time    4.920  ms\n",
       "11  64  bytes  from  172.217.10.4:  icmp_seq  11  ttl  51  time  104.141  ms\n",
       "12  64  bytes  from  172.217.10.4:  icmp_seq  12  ttl  51  time    5.090  ms\n",
       "13  64  bytes  from  172.217.10.4:  icmp_seq  13  ttl  51  time    5.787  ms\n",
       "14  64  bytes  from  172.217.10.4:  icmp_seq  14  ttl  51  time    6.098  ms\n",
       "15  64  bytes  from  172.217.10.4:  icmp_seq  15  ttl  51  time    7.451  ms\n",
       "16  64  bytes  from  172.217.10.4:  icmp_seq  16  ttl  51  time    9.982  ms\n",
       "17  64  bytes  from  172.217.10.4:  icmp_seq  17  ttl  51  time    5.462  ms\n",
       "18  64  bytes  from  172.217.10.4:  icmp_seq  18  ttl  51  time   15.061  ms\n",
       "19  64  bytes  from  172.217.10.4:  icmp_seq  19  ttl  51  time    6.701  ms"
      ]
     },
     "execution_count": 350,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_old = pd.read_table('router_old.txt', delim_whitespace=True, names=(\"A\", \"B\", \"C\", \"D\",\"E\",\"F\",\"G\",\"H\", \"I\", \"Old\", \"L\"))\n",
    "df_old"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>An</th>\n",
       "      <th>Bn</th>\n",
       "      <th>Cn</th>\n",
       "      <th>Dn</th>\n",
       "      <th>En</th>\n",
       "      <th>Fn</th>\n",
       "      <th>Gn</th>\n",
       "      <th>Hn</th>\n",
       "      <th>In</th>\n",
       "      <th>New</th>\n",
       "      <th>Ln</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>0</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.730</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>1</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.673</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>2</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.160</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>3</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.651</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>4</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>3.383</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>5</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.135</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>6</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>6.161</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>7</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>4.919</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>8</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.313</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>9</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>4.059</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>10</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>3.943</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>11</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>4.235</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>12</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>4.183</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>13</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.230</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>14</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>4.255</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>15</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.297</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>16</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.021</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>17</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.268</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>18</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>3.445</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>19</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>4.754</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    An     Bn    Cn              Dn        En  Fn   Gn  Hn    In    New  Ln\n",
       "0   64  bytes  from  172.217.11.36:  icmp_seq   0  ttl  49  time  5.730  ms\n",
       "1   64  bytes  from  172.217.11.36:  icmp_seq   1  ttl  49  time  5.673  ms\n",
       "2   64  bytes  from  172.217.11.36:  icmp_seq   2  ttl  49  time  5.160  ms\n",
       "3   64  bytes  from  172.217.11.36:  icmp_seq   3  ttl  49  time  5.651  ms\n",
       "4   64  bytes  from  172.217.11.36:  icmp_seq   4  ttl  49  time  3.383  ms\n",
       "5   64  bytes  from  172.217.11.36:  icmp_seq   5  ttl  49  time  5.135  ms\n",
       "6   64  bytes  from  172.217.11.36:  icmp_seq   6  ttl  49  time  6.161  ms\n",
       "7   64  bytes  from  172.217.11.36:  icmp_seq   7  ttl  49  time  4.919  ms\n",
       "8   64  bytes  from  172.217.11.36:  icmp_seq   8  ttl  49  time  5.313  ms\n",
       "9   64  bytes  from  172.217.11.36:  icmp_seq   9  ttl  49  time  4.059  ms\n",
       "10  64  bytes  from  172.217.11.36:  icmp_seq  10  ttl  49  time  3.943  ms\n",
       "11  64  bytes  from  172.217.11.36:  icmp_seq  11  ttl  49  time  4.235  ms\n",
       "12  64  bytes  from  172.217.11.36:  icmp_seq  12  ttl  49  time  4.183  ms\n",
       "13  64  bytes  from  172.217.11.36:  icmp_seq  13  ttl  49  time  5.230  ms\n",
       "14  64  bytes  from  172.217.11.36:  icmp_seq  14  ttl  49  time  4.255  ms\n",
       "15  64  bytes  from  172.217.11.36:  icmp_seq  15  ttl  49  time  5.297  ms\n",
       "16  64  bytes  from  172.217.11.36:  icmp_seq  16  ttl  49  time  5.021  ms\n",
       "17  64  bytes  from  172.217.11.36:  icmp_seq  17  ttl  49  time  5.268  ms\n",
       "18  64  bytes  from  172.217.11.36:  icmp_seq  18  ttl  49  time  3.445  ms\n",
       "19  64  bytes  from  172.217.11.36:  icmp_seq  19  ttl  49  time  4.754  ms"
      ]
     },
     "execution_count": 351,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_new = pd.read_table('router_new.txt', delim_whitespace=True, names =(\"An\", \"Bn\", \"Cn\", \"Dn\",\"En\",\"Fn\",\"Gn\",\"Hn\", \"In\", \"New\", \"Ln\"))\n",
    "df_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>F</th>\n",
       "      <th>G</th>\n",
       "      <th>H</th>\n",
       "      <th>I</th>\n",
       "      <th>Old</th>\n",
       "      <th>...</th>\n",
       "      <th>Bn</th>\n",
       "      <th>Cn</th>\n",
       "      <th>Dn</th>\n",
       "      <th>En</th>\n",
       "      <th>Fn</th>\n",
       "      <th>Gn</th>\n",
       "      <th>Hn</th>\n",
       "      <th>In</th>\n",
       "      <th>New</th>\n",
       "      <th>Ln</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>0</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>11.912</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>0</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.730</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>1</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>21.098</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>1</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.673</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>2</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>7.113</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>2</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.160</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>3</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>7.033</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>3</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.651</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>4</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>5.456</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>4</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>3.383</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>5</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>78.941</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>5</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.135</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>6</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>10.592</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>6</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>6.161</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>7</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>7.169</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>7</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>4.919</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>8</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>4.850</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>8</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.313</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>9</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>11.244</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>9</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>4.059</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>10</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>4.920</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>10</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>3.943</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>11</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>104.141</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>11</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>4.235</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>12</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>5.090</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>12</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>4.183</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>13</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>5.787</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>13</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.230</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>14</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>6.098</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>14</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>4.255</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>15</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>7.451</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>15</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.297</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>16</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>9.982</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>16</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.021</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>17</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>5.462</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>17</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>5.268</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>18</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>15.061</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>18</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>3.445</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>64</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.10.4:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>19</td>\n",
       "      <td>ttl</td>\n",
       "      <td>51</td>\n",
       "      <td>time</td>\n",
       "      <td>6.701</td>\n",
       "      <td>...</td>\n",
       "      <td>bytes</td>\n",
       "      <td>from</td>\n",
       "      <td>172.217.11.36:</td>\n",
       "      <td>icmp_seq</td>\n",
       "      <td>19</td>\n",
       "      <td>ttl</td>\n",
       "      <td>49</td>\n",
       "      <td>time</td>\n",
       "      <td>4.754</td>\n",
       "      <td>ms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     A      B     C              D         E   F    G   H     I      Old  ...  \\\n",
       "0   64  bytes  from  172.217.10.4:  icmp_seq   0  ttl  51  time   11.912  ...   \n",
       "1   64  bytes  from  172.217.10.4:  icmp_seq   1  ttl  51  time   21.098  ...   \n",
       "2   64  bytes  from  172.217.10.4:  icmp_seq   2  ttl  51  time    7.113  ...   \n",
       "3   64  bytes  from  172.217.10.4:  icmp_seq   3  ttl  51  time    7.033  ...   \n",
       "4   64  bytes  from  172.217.10.4:  icmp_seq   4  ttl  51  time    5.456  ...   \n",
       "5   64  bytes  from  172.217.10.4:  icmp_seq   5  ttl  51  time   78.941  ...   \n",
       "6   64  bytes  from  172.217.10.4:  icmp_seq   6  ttl  51  time   10.592  ...   \n",
       "7   64  bytes  from  172.217.10.4:  icmp_seq   7  ttl  51  time    7.169  ...   \n",
       "8   64  bytes  from  172.217.10.4:  icmp_seq   8  ttl  51  time    4.850  ...   \n",
       "9   64  bytes  from  172.217.10.4:  icmp_seq   9  ttl  51  time   11.244  ...   \n",
       "10  64  bytes  from  172.217.10.4:  icmp_seq  10  ttl  51  time    4.920  ...   \n",
       "11  64  bytes  from  172.217.10.4:  icmp_seq  11  ttl  51  time  104.141  ...   \n",
       "12  64  bytes  from  172.217.10.4:  icmp_seq  12  ttl  51  time    5.090  ...   \n",
       "13  64  bytes  from  172.217.10.4:  icmp_seq  13  ttl  51  time    5.787  ...   \n",
       "14  64  bytes  from  172.217.10.4:  icmp_seq  14  ttl  51  time    6.098  ...   \n",
       "15  64  bytes  from  172.217.10.4:  icmp_seq  15  ttl  51  time    7.451  ...   \n",
       "16  64  bytes  from  172.217.10.4:  icmp_seq  16  ttl  51  time    9.982  ...   \n",
       "17  64  bytes  from  172.217.10.4:  icmp_seq  17  ttl  51  time    5.462  ...   \n",
       "18  64  bytes  from  172.217.10.4:  icmp_seq  18  ttl  51  time   15.061  ...   \n",
       "19  64  bytes  from  172.217.10.4:  icmp_seq  19  ttl  51  time    6.701  ...   \n",
       "\n",
       "       Bn    Cn              Dn        En  Fn   Gn  Hn    In    New  Ln  \n",
       "0   bytes  from  172.217.11.36:  icmp_seq   0  ttl  49  time  5.730  ms  \n",
       "1   bytes  from  172.217.11.36:  icmp_seq   1  ttl  49  time  5.673  ms  \n",
       "2   bytes  from  172.217.11.36:  icmp_seq   2  ttl  49  time  5.160  ms  \n",
       "3   bytes  from  172.217.11.36:  icmp_seq   3  ttl  49  time  5.651  ms  \n",
       "4   bytes  from  172.217.11.36:  icmp_seq   4  ttl  49  time  3.383  ms  \n",
       "5   bytes  from  172.217.11.36:  icmp_seq   5  ttl  49  time  5.135  ms  \n",
       "6   bytes  from  172.217.11.36:  icmp_seq   6  ttl  49  time  6.161  ms  \n",
       "7   bytes  from  172.217.11.36:  icmp_seq   7  ttl  49  time  4.919  ms  \n",
       "8   bytes  from  172.217.11.36:  icmp_seq   8  ttl  49  time  5.313  ms  \n",
       "9   bytes  from  172.217.11.36:  icmp_seq   9  ttl  49  time  4.059  ms  \n",
       "10  bytes  from  172.217.11.36:  icmp_seq  10  ttl  49  time  3.943  ms  \n",
       "11  bytes  from  172.217.11.36:  icmp_seq  11  ttl  49  time  4.235  ms  \n",
       "12  bytes  from  172.217.11.36:  icmp_seq  12  ttl  49  time  4.183  ms  \n",
       "13  bytes  from  172.217.11.36:  icmp_seq  13  ttl  49  time  5.230  ms  \n",
       "14  bytes  from  172.217.11.36:  icmp_seq  14  ttl  49  time  4.255  ms  \n",
       "15  bytes  from  172.217.11.36:  icmp_seq  15  ttl  49  time  5.297  ms  \n",
       "16  bytes  from  172.217.11.36:  icmp_seq  16  ttl  49  time  5.021  ms  \n",
       "17  bytes  from  172.217.11.36:  icmp_seq  17  ttl  49  time  5.268  ms  \n",
       "18  bytes  from  172.217.11.36:  icmp_seq  18  ttl  49  time  3.445  ms  \n",
       "19  bytes  from  172.217.11.36:  icmp_seq  19  ttl  49  time  4.754  ms  \n",
       "\n",
       "[20 rows x 22 columns]"
      ]
     },
     "execution_count": 352,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#routers_df = df_old.append(df_new, sort=True)\n",
    "\n",
    "routers_df = pd.concat([df_old, df_new], axis=1)\n",
    "routers_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 361,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Old</th>\n",
       "      <th>New</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>11.912</td>\n",
       "      <td>5.730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>21.098</td>\n",
       "      <td>5.673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>7.113</td>\n",
       "      <td>5.160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>7.033</td>\n",
       "      <td>5.651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>5.456</td>\n",
       "      <td>3.383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>78.941</td>\n",
       "      <td>5.135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>10.592</td>\n",
       "      <td>6.161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>7.169</td>\n",
       "      <td>4.919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>4.850</td>\n",
       "      <td>5.313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>11.244</td>\n",
       "      <td>4.059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>4.920</td>\n",
       "      <td>3.943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>104.141</td>\n",
       "      <td>4.235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>5.090</td>\n",
       "      <td>4.183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>5.787</td>\n",
       "      <td>5.230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>6.098</td>\n",
       "      <td>4.255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>7.451</td>\n",
       "      <td>5.297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>9.982</td>\n",
       "      <td>5.021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>5.462</td>\n",
       "      <td>5.268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>15.061</td>\n",
       "      <td>3.445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>6.701</td>\n",
       "      <td>4.754</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Old    New\n",
       "0    11.912  5.730\n",
       "1    21.098  5.673\n",
       "2     7.113  5.160\n",
       "3     7.033  5.651\n",
       "4     5.456  3.383\n",
       "5    78.941  5.135\n",
       "6    10.592  6.161\n",
       "7     7.169  4.919\n",
       "8     4.850  5.313\n",
       "9    11.244  4.059\n",
       "10    4.920  3.943\n",
       "11  104.141  4.235\n",
       "12    5.090  4.183\n",
       "13    5.787  5.230\n",
       "14    6.098  4.255\n",
       "15    7.451  5.297\n",
       "16    9.982  5.021\n",
       "17    5.462  5.268\n",
       "18   15.061  3.445\n",
       "19    6.701  4.754"
      ]
     },
     "execution_count": 361,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_fig = routers_df[[\"Old\", \"New\"]]\n",
    "df_fig "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 366,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "plt.style.use('seaborn-whitegrid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 378,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'Runtime (secs)')"
      ]
     },
     "execution_count": 378,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEBCAYAAAB/rs7oAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3deXhTVfoH8O/N2jZJlzSFUtoChVaKCgiVRaG4gLjhigLKpqOC+lPLoIIgy4wLOow4MygDuCKLIgruo46MgrJqZS0FhLZAQylNm7ZJmj3398ftTdM1+9a+n+fJ0/RmuW8v4b0n557zHoZlWRaEEEK6DEG4AyCEEBJalPgJIaSLocRPCCFdDCV+QgjpYijxE0JIF0OJnxBCuhhRuANwp7CwMNwhEEJIVBo6dGib2yM+8QPtBx+JiouLkZubG+4wPEbxBl+0xUzxBleo4u2o0UxdPYQQ0sVQ4ieEkC6GEj8hhHQxlPgJIaSLocRPCCFdDCV+QgjpYijxE0JIF0OJn5AI9+4vpZj+7v5wh0EC4Ny5c3jllVcwbdo0TJ48GUuXLoVer8f8+fOxc+fOZs81m8247rrrghJHVEzgIqQr++1MDfaergbLsmAYJtzhEB+ZTCY89thjeOihh3D77bcDALZt24a5c+ciKSkppLFQ4ickwml0FljsDtQbbUiIE4c7nKj3aWE5Pv7tXEDf8968DNw9NL3D5/z000+48sorkZOT49x255134sMPP0RCQgIAwGAw4Omnn0Z9fT0yMzMDGqMr6uohJMJp9GYAQFXjTxKdzp0712YyT09Px2+//QaA+waQk5ODjRs3YvLkyUGLhVr8hES4Kp3Z+bNfN3mYo4l+dw9Nd9s6D4bu3bvj8OHDGD58eLPtZWVlyMvLAwD88ccfGD16NABg0KBBEImCk6KpxU9IBDNZ7dCZbQCaWv4kOl1//fXYvXs3Tp486dy2ZcsWKJVKCARcKs7KysLBgwcBAMeOHYPNZgtKLEFL/IcOHcK0adMAAGfOnMGUKVNw3333YcmSJXA4HACAN954AxMnTsTkyZNx+PDhYIVCSNRyTfaU+KObTCbD6tWrsWXLFkyePBn33HMPDh06hBUrVjifc//996OyshJTpkzBxo0bIRYH55pOUL5HvPXWW/jiiy8QGxsLAFi2bBkKCgowfPhwLF68GNu3b0daWhr279+PLVu2oKKiAk888QQ+/fTTYIRDSNTS6C0u9ynxR7vMzEwsXLiwVVnmV155xXl/+fLlQY8jKC3+zMxMrFy50vl7UVERhg0bBgDIz8/H7t27UVhYiFGjRoFhGKSlpcFut6OmpiYY4RAStTS6pmRfpaPETwIjKC3+8ePHo7y83Pm76/hjmUwGnU4HvV6PxMRE53P47UqlstX7FRcXByPMoDCZTBRvEEVbvIB/MR/5ox4AEC8VoOxCTUj+9mg7xhSv90Iyqoe/cAFw41Tj4+Mhl8thMBiabVcoFG2+nlbXCR6KN/j8ifmH838A0ODSnkkwWGwh+duj7RhTvG0L+wpcAwYMwL59+wAAO3fuRF5eHoYMGYJffvkFDocD58+fh8PhaLO1T0hXptGbER8jQlpibLNuH0L8EZIW/7x587Bo0SKsWLECWVlZGD9+PIRCIfLy8jBp0iQ4HA4sXrw4FKEQElU0egtUCilUCgk0eguVbSABEbTEn56ejo8//hgA0KdPH2zYsKHVc5544gk88cQTwQqBkKhXpTdDJZciRS6lsg0kYGgCFyERTKM3I0UuRYpCCoDKNkSzffv2IS8vD1VVVc5tf//737F169aQx0KJn5AIptGZoZJLoJJziZ/G8kc3sViMlStXgmXZsMZBtXoIiVAmqx31JhtSFFJK/IF08EPgQOuuZ79cMRUYPMXt00aMGIHa2lps3LgRU6dOdW5fv349vvrqKzAMg5tvvhkTJkzAzJkz8fnnn+PAgQOYNWsW9uzZg6qqKixcuBDvvPOOX+FSi5+QCFVt4GbtquRSqOQSAKCRPZ3A7Nmz8f7776OsrAwAYDQa8c0332DTpk3YtGkTfvjhB2i1WiQmJqKiogI///wzUlNTUVRUhO3bt2Ps2LF+x0AtfkIiFJ/kVXIpkuIkEAoY6uMPhMFTPGqdB0t8fDwWLFiA+fPnY8iQIWhoaMD58+cxc+ZMAEBdXR3Onj2LcePGYceOHThw4AAefvhh7Nq1CwcOHMBLL73kdwzU4ickQvHdOiqFFAIBg2SZBBqdxc2rSDS47rrr0KdPH2zbtg0SiQT9+vXDBx98gPXr1+Ouu+5CTk4Oxo4di6+++gpyuRz5+fn44YcfYLFYkJKS4vf+qcVPSIRyJv7Gbh6VXEp9/J3IwoULsXfvXigUCowcORJTpkyBxWLBwIED0b17dwiFQpjNZowYMQIJCQkQiUS45pprArJvSvyERCi+Mid/YVeloMQfzYYPH47hw4c76/TI5XL8+OOPzscfeuihVq/ZsmWL8/7mzZsDFgt19RASoap0ZihiRIgRCwFwLX+q0EkCgRI/IRGqqnHyFi9FLnWWbSDEH5T4CYlQ3OQtl8SvaCzbYArOcnyk66DET0iE0ujNUCkkzt9pEhcJFEr8hEQojd7SrMXvTPzUz0/8RImfkAhkttlRZ7Q26+PnW/80iYv4ixI/IRGomh/KqWh+cRegFj/xHyV+QiJQ0+StpsTPl23gx/cT4itK/IREoJazdgFAIGCglEno4i7xGyV+QiIQX5PHtcXP/06Jn/iLEj8hEYi/gJuiaJn4afYu8R8lfkIiUJXODIW0qVwDL0UhpT5+4jdK/IREIG7ylrTV9hS5FFV6M5VtIH6hxE9IBNLozc0u7PJUciksNgd0ZirbQHxHiZ+QCNRy1i7POYmL+vmJHyjxExKBuBZ/G4mfJnGRAKDET0iEsdgcqG2wtpn4+VE+dIGX+IMSPyERptrQ9lBOgCp0ksCgxE9IhGmavNX64m5SnAQChhI/8Q8lfkIijLNcQxstfqGAgVImpYu7xC+U+AmJMM5Zu2308QP8JC5K/MR3lPgJiTBtVeZ0pZJLUEUXd4kfKPETEmE0OgvkUhFiJcI2H0+RS2k4J/GLKFQ7slqtmD9/PtRqNQQCAV544QWIRCLMnz8fDMMgOzsbS5YsgUBA5yLStVW1M2uXp2rs6mFZFgzDhDAy0lmELMvu2LEDNpsNH330ER5//HH84x//wLJly1BQUIBNmzaBZVls3749VOEQErE0urYnb/FUcgnMVLaB+CFkib9Pnz6w2+1wOBzQ6/UQiUQoKirCsGHDAAD5+fnYvXt3qMIhJGK1N2uX55zERd09xEch6+qJi4uDWq3GTTfdBK1Wi9WrV+PXX391flWVyWTQ6XRtvra4uDhUYfrNZDJRvEEUbfEC3sdcWdeAnCRBu69pqGkAABQWnYS5e2xAYnQVbceY4vVeyBL/+++/j1GjRmHu3LmoqKjAjBkzYLVanY8bDAbEx8e3+drc3NxQhem34uLiqIlXb7bhn1/+ivl3DYZQEB19xdF0fHnexGy1O1BvLkF2Rnfk5ua0/aTEeuC/FyBL7oHc3B4BjJQTbceY4m1bYWFhu4+FrKsnPj4eCoUCAJCQkACbzYYBAwZg3759AICdO3ciLy8vVOEQAN8XXcBbv9XgwFltuEMhjaobh2m2Va6Bx3cD0SQu4quQJf6ZM2eiqKgI9913H2bMmIE5c+Zg8eLFWLlyJSZNmgSr1Yrx48eHKhwCoFxrBACc0zaEORLCczeGHwCUMirbQPwTsq4emUyGf/7zn622b9iwIVQhkBbUjYn/bLUxzJEQXpUHiZ8v20CJn/iKBs13YeW1XEv/bA21+CMFP1KnvXINPG7RdZq9S3xDib8LU1NXT8Th6+zzK221h+r1EH9Q4u+iHA4W52tNAIBz1OKPGBq9GXESIeIkHffCquRUoZP4jhJ/F6XRm2GxO5AYI8SFehPMNnu4QyLgRup0NKKHl+JStoEQb1Hi76LONXbzDEyNAcs2dfuQ8HI3a5fHl23QU9kG4gNK/F2UupZL9INSuZmfdIE3MmjcFGjjNS3BSBd4ifco8XdRapcWP9D0DYCEl0Zv8bDFT2vvEt9R4u+iyrUNSIwTIy1eDIlIQBd4I4DN7oC2wbvETxd4iS8o8XdR6loj0pNiIWAYpCfF4mw1Jf5wqzFYwLJtr7XbkrNCJ7X4iQ8o8XdRaq0RPRO5/v1MZRyN5Y8AFz2cvAW4lG2gFj/xASX+LohlWZRrjeiZGAeAS/xnqxtoaGCY8a33FDeTtwC+bAOtvUt8Q4m/C9I2WGG02pGexLX4M5LioDPbUGe0unklCSbnrF0PWvz886iPn/iCEn8XxI/o6cknfiXX8j9XQyN7wsmTypyuqGwD8RUl/i5I3ViczbWPH6Cx/OGm0ZkRKxZCJvWsaK5KTomf+MajT9jJkydRW1uL5ORk9O3bN9gxkSDj6/CnJ8Wioh7IUNIkrkig0XtWroGnkkucZRv4JUwJ8US7id9isWDt2rX49ttvkZycDJVKhfr6elRWVuLmm2/GzJkzERMTE8pYSYCUa42QS0VIiBWjAoAiRoykODGN7AmzKg9n7fJUcilMVgcMFjvkHn5LIAToIPEvXrwYEyZMwGOPPQaBoKlHiGVZ7Ny5E4sXL8bf/va3kARJAktdyw3ldG0lZirjaBJXmGl0FvRKjvP4+a6TuCjxE2+0+2l55ZVX2tzOMAzGjBmDMWPGBC0oElzlWqPzwi4vXRmHInVdmCIiANfVM7R3ksfPd53E1UclC1ZYpBNye3H3jz/+wIEDB3Do0CHMmDEDu3fvDkVcJIjU2gbnUE5epjIO5Voj7A4ayx8ONrsDNR6Wa+A56/XQkE7iJbeJf8mSJZBIJPj3v/+NOXPm4M033wxFXCRIdCYr6k0254geXqYyDjYHi4o6GtIZDjUNXLmGFG/6+BsnetHIHuItt4lfJBIhOzsbVqsVgwcPht1OC3ZEM74cc8uunowkGssfTvxELG9G9SjjJGAY0Oxd4jW3iZ9hGMydOxf5+fn45ptvEBsb6+4lJIKVNyb2tlr8AC3DGC7eztoFAJFQAGWchGbvEq+5HQrw+uuv48iRI8jPz8f+/fvx+uuvhyIuEiR8iz89qfnokR6JMRAKGBrLHyZ8P703iR+g2bvEN25b/AcPHsTBgwfBMAzWrl2Lo0ePhiIuEiTqWiOkIkGr8eJioQA9EmJoLH+YOMs1eNHVA9DsXeIbt4l/5cqVmDp1KgDgH//4B954442gB0WCp1zb0GoMPy9TGUct/jDR6M2IEQsgkwi9eh0/e5cQb3h0cTc5ORkAoFAomk3mItFH3cYYfl5GEk3iChd+yUVvSy/wFTqppDbxhts+/oEDB2Lu3LkYPHgwDh8+jAEDBoQiLhIk6lojBqTFt/lYZnIcNHoLGiw2xEloJmgoVem8q9PDS1FQ2QbiPbfN9+effx433XQTTCYTbrrpJjz//POhiIsEgdFih0ZvaTWih0flmcNHozd7fWEXoElcxDduE7/BYMCRI0dQWloKm82GM2fOhCIuEgTtjeHnZTRup+6e0PM58dPau8QHbhP/ggULkJGRgbKyMqhUKixcuDAUcZEgaG8oJ4/q8oeH3cGixmDxatYujx+dRYmfeMNt4q+trcXEiRMhEokwZMgQuogUxZwrb7XT1aOUSSCTCCnxh1iNwQIH6/1QTqBpYXaaxEW84dHVoNOnTwMALly44NeonjVr1uB///sfrFYrpkyZgmHDhmH+/PlgGAbZ2dlYsmQJjRoKonJtA0QCBt3j215HgWEYZCjjUE5j+UPK2yUXXSllVLaBeM+ji7sLFizAsWPH8OSTT2L+/Pk+7Wjfvn04cOAAPvzwQ6xfvx4XLlzAsmXLUFBQgE2bNoFlWWzfvt2n9yaeUdcanTN025NBY/lDzpc6PTy+bAN19RBvuG3x5+TkYPPmzQCAiooK9OjRw6cd/fLLL8jJycHjjz8OvV6PZ599Fh9//DGGDRsGAMjPz8euXbswbtw4n96fuKfWGtvt5uFlKuPwyx8aWs4vhPxp8fOvo1E9xBtuE/8HH3yAmJgY1NfXY+vWrRg9ejSee+45r3ek1Wpx/vx5rF69GuXl5Xj00UebJReZTAadTtfma4uLi73eX7iYTKaIjbe0qh5X9IhtFl/LeMWWehitduw5cBRJsZE3LjySj2973MVcdLoWAFCtLoOpyvuuzliBDeeqagN2XKLtGFO83nP7P/vrr7/G+vXr8dBDD+Hrr7/GjBkzfNpRYmIisrKyIJFIkJWVBalUigsXLjgfNxgMiI9ve2JRbm6uT/sMh+Li4oiM12JzoMZYggG9eyA3N8e5vWW8FUwlVu+vhjQ5Hbm9PF8NKlQi9fh2xF3MwtJiSEW1GDpwgE/fsnodMuH3s9qAHZdoO8YUb9sKCwvbfcyjssxVVVVQqVRgGAZ1db4tzzd06FD8/PPPYFkWlZWVMBqNGDlyJPbt2wcA2LlzJ/Ly8nx6b+LehToTWBatVt5qicozh55GZ/apXAMvRS6FRkcXd4nn3Lb4hw8fjqlTp+K1117Dyy+/jBtuuMGnHV177bX49ddfMXHiRLAsi8WLFyM9PR2LFi3CihUrkJWVhfHjx/v03sS98loukae76eNPT6LEH2pVet/KNfBUCimMVjsMZhtkVLaBeMDtp2TOnDmYM2cOAOCyyy6DROL9JBPes88+22rbhg0bfH4/4rlybcezdnkxYiG6KaQ0sieEqnTmdifVecJZtkFvpsRPPNJuV8+zzz6Ln376qdlSixKJBA6HAz/88AOefvrpkARIAkOtNYJhgB4J7ldQoyGdoaXRW5Ci8L1Bxc/epUlcxFPtNg9efPFFrFu3Dq+99hoUCgVUKhXq6upQU1ODCRMm4KWXXgplnMRP6lojuitiIBG5HzWSqYzD/tKaEERFuHINvtXp4aVQvR7ipXYTv0QiwcMPP4yHH34YZWVl0Gq1SE5ORmZmZijjIwFSrm1w283Dy1DG4bODalhsDo9OFMR32obGcg3+JH6+bAPN3iUe8qhDsHfv3ujdu3eQQyHBpK414ooMz4ZnZiTFgmWB87VG9FbJghxZ1+bv5C2gqWwDTeIinqLmXBdgd7CoqDW5HcrJoyqdoeNPuQaeSChAEpVtIF6gxN8FXNSZYHOwHnf1ZCZT4g+Vpha/7xd3+dfTxV3iKbddPSdPnsTSpUuh0+kwYcIEZGdn49prrw1FbCRAyt2UY26puyIGEqEA56hKZ9DxE698KcnsKkUhpRY/8ZjbFv9LL72EZcuWITExERMnTsTKlStDERcJIL4Ov6djxQUCBulJsTSJKwQ0ejMkIgEUfo6/V8ml0NDFXeIhj7p6evXqBYZhoFQqIZPRxb5o41xy0cMWP0Bj+UOlSm9Gih/lGnhc4qcWP/GM28SfkJCAjz76CEajEV9//XW7hdRI5CrXNiBZJkGsROjxazKUsbToegho9Ba/+/cBLvE3WLiyDYS44zbxv/zyyygvL0dSUhKOHj1KE7eiULnW6PGIHl6mMg51RivqGqxBiooA3Kgef0b08GjtXeINtx2LsbGxuPXWW2GxcP2HZ8+eRWJiYtADI4GjrjWif6rCq9c4q3RqG5AQlxCMsAi4RD0o3f/j6zp7t1cydceSjrlN/I888ggsFgsSEhKcC6e88cYboYiNBADLslBrjbi+fzevXudapfOynpT4g8HhYFFjsPg1eYunci66Thd4iXtuE7/ZbKYKmlFMo7fAbHN4dWEXoLH8oaBtsMDuYAPSx0/1eog33Cb+vLw8/Pzzz+jbt69zW1paWlCDIoHDj+jxtuxvfIwYCbFiSvxBxA+/9HcMP8CVbQCoQifxjNvEX11djZdfftk5modhGHz00UdBD4wEhtrDOvxtyVTG4ZyWRvYESyDq9PDEQgGUMirbQDzjNvGXlpbiP//5TyhiIUFQ3jj71tfEf6yiPtAhkUaBqNPjSiWnxE8843Y4Z05ODg4ePAiLxeK8keihrjUiPkaE+Bix169NV8ZCrTXC7mCDEBkJZIuffx+avUs84bbF/+uvv+Knn35y/s4wDLZv3x7MmEgAqbVG9PRxWb9MZRwsdgcq601I8/LiMHGvSm+GRChAfExglktUyaU4VF4bkPcinZvbT9yXX34ZijhIkJRrjchQ+p74AW5kDyX+wNPouFm7/pZr4KnkUrq4SzzSbuL/61//isWLF2PSpEmtPph0cTc6sCwLda0RI/sm+/T6DJex/COyfHsP0j6N3hyQET28FAVXtqHBYkOchBZdJ+1r99Px2GOPAQBeffVViMVN/cN1dXXBj4oERL3RBr3Z5nW5Bl5aYiwEDKhKZ5BU6czokRATsPdzlm3QWZCZTImftK/di7ssy6K0tBTPPvssrFYrLBYLTCYTFi9eHMr4iB/KaxtH9PjYTSMRCdAjIZbG8geJRu/fIust8d8eqmhkD3Gj3WbBoUOHsG7dOpSWlmLRokUAAIFAgFGjRoUsOOKfcj/G8PMylLE0lj8IHA4W1QYLVAr/Z+3y+EXXaUgncafdxD927FiMHTsWO3bswJgxY0IZEwkQbxdgaUumMg4/nqgKVEikUa3R2liuIbB9/ADN3iXuue0I7NatG5YuXQqzuenDtGzZsqAGRQJDXWtErFiIpDjvx/DzMpVxqNKZYbTYvarnTzoW6DH8QFPZBmrxE3fcJv758+dj6tSpSE1NDUU8JIDKtQ3omRTr13BBfihoubYB2d29K+1M2qfRBT7xi4UCJMWJKfETt9wmfpVKhXvuuScUsZAAU9cafb6wy8twGctPiT9w+AuwgSrXwFPJpc4F3Alpj9vE37NnT6xduxa5ubnOliNd4I0Oaq0Rg9L9WzSHH8tPI3sCy1mnJ4AtfqBxEhe1+IkbbhO/1WpFaWkpSktLndso8Uc+g9kGbYPVrxE9ADc2PFYspPV3A0yjt3DlGmIDO94+RUFlG4h7bj91dCE3OvF1+P3t6mEYBpnKOGrxB5hGb0ZyAMs18LiuHmrxk465Tfyurfva2lpkZGT4Vaa5uroad911F959912IRCLMnz8fDMMgOzsbS5YsgUDgtmAo8UAghnLyMpSxzvLOJDACPXmLp1JIYLDYaRQW6ZDbLPvLL784b9999x0GDx7s886sVisWL16MmBhumvqyZctQUFCATZs2gWVZqvoZQOXOlbf8L66W0djiZ1kqzxwoXOIP3OQtnoomcREPeNW87tmzJ0pKSnze2auvvorJkyejWzdu4e+ioiIMGzYMAJCfn4/du3f7/N6kuXJtAyRCQUAuHmYq49BgsaPaQKNFAqVKZw74iB6g6WLxReruIR1w29Xz5z//2dkPefHiRSQn+1alcevWrVAqlRg9ejTWrl0LgKsHxL+3TCaDTqfz6b1Ja2qtET0SYyAQ+N+H7FqlMxjdE12Nw8GiWm8JyrGkRdeJJ9wm/smTJzvvS6VSXH755T7t6NNPPwXDMNizZw+Ki4sxb9481NTUOB83GAzOdX1bKi4u9mmf4WAymSIi3lMVNUiSCNzG4km8tlqupb/3yCnEGOQBi9EXkXJ8vdEy5nqTHTYHC7uhNuB/i9ZgAwAc/eMMMgVan94j2o4xxeu9dhO/zWbD//73P8THx2PEiBEAAI1Ggz//+c/4xz/+4fWONm7c6Lw/bdo0LF26FMuXL8e+ffswfPhw7Ny507mflnJzc73eX7gUFxdHRLzVW9W49pJkt7F4Em9vix34ohz22ETk5mYHMkyvRcrx9UbLmP+o1AE4g0v7ZSI3Ny2g+7LaHcAnZyGSK33+t4q2Y0zxtq2wsLDdx9pN/E8//TSEQiGqqqpw6tQppKenY+HChZg+fXrAAps3bx4WLVqEFStWICsrC+PHjw/Ye3dlJqsdVTozeib6P6IHAGIlQqjkUhrLHyBVzjo9gb+4KxYKkEhlG4gb7Sb+s2fPYuvWrbBYLLj77rshFovxwQcfoG/fvn7vdP369c77GzZs8Pv9SHMVdSYAgRnRw8tUUl3+QOEXRA/0rF1eCi3BSNxoN/HL5VxfrkQigcPhwLvvvovERP+m/5PQUAegDn9Lmco4/FrmW58xac5ZriEIo3qAxklc1OInHfBoOGdycjIl/SjCT7byd9auqwxlHCrqjFwfMvGLRm+GWMggIdb3ctkdUSko8ZOOtdviP3XqFObOnQuWZZ33ea+99lpIgiO+UdcaIWCA1ACu55qhjIODBc7XGtErWRaw9+2KNDozkmXSgJdr4KnkEmd3EiFtaTfxu47ccR3SSSKfWmtEj4RYiIWBK3/RNJafEr+/NHpzQJdcbEkll0JvtlHZBtKudhM/P6OWRJ9yrf91+FvKTKbyzIGiCdLkLZ7rJC5+PQVCXFFFtE5IXWsM6IVdAEiNj4FYyFDiD4AqnTloI3qAptFCVJeftIcSfydjsztwod4U0KGcACAUMOiZGItzVKXTLyzLotpghipII3oAl0JtNKSTtIMSfydTUWeC3cEGvKsH4C7wnqMWv1/qjFZY7WxQu3r46wd0gZe0hxJ/J+NcgCXALX4AtCBLAGiCOGuXlyxr7OqhFj9pByX+TsY5eStILf7aBivqTdaAv3dXUaUL7qxdAJCIqGwD6Rgl/k6Gb/GnBSHxZyqbyjMT3zhb/EHs4wdo9i7pGCX+TqZc24AUhRQx4sCP36bE7z9nuYYgr2vATeKixE/aRom/k1HXBn4MP891EhfxjUZvhkgQvHINPK7FTxd3Sdso8Xcyaq0x4EM5eQlxYsTHiOgCrx80ejOS5ZKArIzWkRQFVegk7aPE34k4HCzO15qCMqKHl5lMI3v8EexZuzy+bIPJag/6vkj0ocTfiVTpzbDYHUgPUlcPwHX30CQu32n05pAkfufsXWr1kzZQ4u9EyoNQh7+lTGUcymuMcDjYoO2jM9PoQpP4myZxUeInrVHi70T4oZzpScErzJWujIPF7sBFakl6jWVZaPSWoC3A4kpFLX7SAUr8nUgwFmBpiR/SSf383qs32mCxO4I6a5fXVKGTRvaQ1ijxdyJqrRGJcWLIpO1W2/YbJX7f8dUyQ9Hi58s2UFcPaQsl/k5EXRu8oZy8tMQYMAxN4ocCk44AACAASURBVPJFU52e4Cd+iUiAhFgq20DaRom/EwnGAiwtSUVC9IiPocTvg1Amfm4/NHuXtI0SfyfBsizUWiN6JgZ/xaUMqtLpE74+fij6+Ln90CQu0jZK/J2EtsEKo9Ue1KGcvAwljeX3RZXeDKGAQVJcaBJ/ioLKNpC2UeLvJPgRPcHu4we4C7yV9WaaFeoljc6CZFnwyzXwVHIprcJF2kSJv5MIZh3+lviRPeXU6vdKqGbt8lIUUuiobANpAyX+TqJp8lYounq4fVCVTu9o9MFda7cl/loCXeAlLVHi7yTKtUbIpaKgl/sFuD5+gMbye4sr0Baa/n2gab4AXeAlLVHi7yT4oZwME/z+4xS5FDFiASV+L7AsiyqdOSSTt3h8txJd4CUtUeLvJNS1xpCM6AEAhmG4Kp2U+D1Wb+LKNQR75S1XTYmfWvykOUr8nYRa2xCSC7u8TBrL75VQT94CgGS+j5+6ekgLXSLx2x1spy4jXG+yot5kC8mFXV6GMg7lWiNYtvMe10BqmrwVusQvFQkRHyNy1ggihBe8al4tWK1WLFiwAGq1GhaLBY8++ij69euH+fPng2EYZGdnY8mSJRAIAnsuqqw34f6396GXMg5vz8gLSR94qKlDUIe/pQxlHPRmG7QNVihlobtgGa34fna+Tn6ocJO4KPGT5kLW4v/iiy+QmJiITZs24a233sILL7yAZcuWoaCgAJs2bQLLsti+fXtA91mlM+O+t/aiVGPA9uMX8UlheUDfP1KEcgw/j6p0eiccXT38/jQ6urhLmgtZ4r/xxhvx1FNPOX8XCoUoKirCsGHDAAD5+fnYvXt3wPZXrTfj/rf34nytCRsfGo4reyfhxa+LO+XQtlAswNJS01h+SvyeqNKFtlwDT0UtftKGkHX1yGQyAIBer8eTTz6JgoICvPrqq86uF5lMBp1O1+Zri4uLvdqXzmzH/O8qUF5vxV+uT0WC+SIeHizDY19oMXfjHjw3prt/f0wHTCaT1/H669CpakiEDKrOnYbGy64sX+M1Wh0AgMITZegnqfP69b4Kx/H1l8lkwqlyHeKlApw8cTyk+xZaDKisM3p1zKLtGFO83gtZ4geAiooKPP7447jvvvswYcIELF++3PmYwWBAfHx8m6/Lzc31eB91Rivuf3svynU2vD3jSuTnpHDvAeApQxxe++9JTMtXYtyA4CT/4uJir+INBNPvhUhPsmHAgAFev9afeFVfnIdJKA/p3xuO4+uv4uJi2EQ2pCYyIY/9kgun8MXxE+jTLwcxYqFHr4m2Y0zxtq2wsLDdx0LW1aPRaPDggw/imWeewcSJEwEAAwYMwL59+wAAO3fuRF5enl/70JmsmP7ufpy4oMOaqUOdSZ83a0xf9E9VYNFnR1Fvsvq1r0ii1oZuDL+r9CSq0umpqhDP2uVR2QbSlpAl/tWrV6O+vh6rVq3CtGnTMG3aNBQUFGDlypWYNGkSrFYrxo8f7/P76802zHzvVxSp6/DmfUNwbf9urZ4jEQnwyt0DcVFnwt++De1X7mAq1wZ/5a220Fh+z2l05pBO3uLR7F3SlpB19Tz//PN4/vnnW23fsGGD3+/dYLHhwfd/xcFztXhjyhW44dLUdp87OCMRD1zdB+/8UorbBvXEsD5Kv/cfTkaLHdUGS0hH9PAylXH4+kgFbHYHRMIuMSXEJyzLhrxAG8+Z+DvhoAbiu6j/32qy2vHQut/wW1kNXp80GDdd3sPta+bekIP0pFjM33o46kvW8iN6wtHVk6GMhd3BoqLOFPJ9R5MGKwuzLbTlGnj8yYa6eoirqE78JqsdD3/wG/aUVOO1ewfhtkFpHr0uTiLCy3dejpIqA9788VSQowyupgVYQjeUk0dVOj2jNdoAhH7yFtDUx98ZhzET30Vt4jfb7Hh0QyF+/kODV+8eiDuvSPfq9fk5KbhrSE/8+6fTKK6oD1KUweds8YepqwegxO9OrYn7VhnqyVtAU9kGavETV1GZ+K12B/5v0wH8eKIKL995Oe7Ny/DpfRbdMgAJsWLM//Qw7FFay0etNUIkYNA9Pibk++6REAuRgKFJXG5ojeFL/AA/iYsu7kYDlmVRqjEEvQZW1CV+m92Bpz46gP8eq8Rfb78U9w3P9Pm9kmQSLLntUhwqr8N7u0oDGGXoqGuN6JEYA2GI1nF1JRQw6JkUSy1+N8LZ4uf3S4XaosPfvz+Ba//+E57/7GhQG6NRlfjtDhZzPj6Eb45cwPO35GL6yN5+v+eEgT1wXf9ueO37k1HZcuUXYAmXTGUczmlpCcaOaI12CBiErZhdCi26HhXW7DiNN388jQE94rFx31kUbD4Ii80RlH1FTeK3O1g8s+UQvjx0HvNv6o+HRmcF5H0ZhsGLd1wGAQMs2HYk6soMq7VG9EwM/YVdXjotyOKW1miHUiYNy7cygKvQSS3+yPbh/rNY9p/jmDAoDV8+MQrzb+qPLw+dxyPrf4PREviRh1GR+B0OFs9tPYytB9SYOy4Hs8f0Dej7pyXGYt5N/fHzHxps/V0d0PcOJovNgUqdKSxDOXmZyjjUGCzQm21hiyHS1ZrsYZm1y1PJJdCZbFE/dLmz+vLQeSzYdgTXXpKCFfcOglDAYPaYvlh21+XYcbIK09/dhzpjYCsNREXif/7zo/j4t3I8eX02nrg+Oyj7mDq8F4b2SsILXx+LmhEQFXVGsCzCMmuXR1U63dMa7SFda7cl/tpCtYEu8EaaH49fxJzNB3FlbyVW3T8UYpeJkFOGZeKNKUNw8FwtpqzdG9AhuVGR+DftO4tHr+mLOWODk/QBQCBg8Mpdl6PBbMdfvjwWtP0EEl+HPz3MffwADensCNfiD3/ip37+yLKvpBqzNxSifw8F3pmRh1hJ6yJ6twzsgbdnXIlSjQH3rN7tnLfjr6hI/A+N6oNnx18S9NWzsrsr8Pi1/fDlofPYXlwZ1H0FQnkYZ+3y+MRPLf62sSyLWmN4u3r4bxs0iStyHFXX4aF1vyE9KRbrHhgGRYy43eeOyUnBhoeGocZgwcR/78Gpi22Xr/dGVCT+hbfkhmzJxEev6Yuc7nI8/9lR6CK8gqdaawTDcOPpwyUhVgyFVNSpEr/OZMXiz4/iX9v/gNXu36gKvdkGs50Nb1cPlW2IKOfqLJj+7n7Ex4qx4aHhSPbg2+DQXkpsnjUSNgeLe1bvweHyWr9iiIrEH8p1cvkKnhfqTVj+3YmQ7dcX5VojuitiIBGF75+RYRhkKONwrKI+KKMPQq3wTA1u/tfPWL/3DFb89yQm/ns3Sqr0Pr+fc63dMHb1JMuoNHOkKNc2YMH3FRAwDDY8NNyrRltuj3h8MnskZFIRpqzdiz2nq32OIyoSf6gNyUzCzKt6Y/3eM/itrCbc4bRLXdsQ1m4e3tX9kvFrmRbDXvoBz209gt/PaqNuWKzN7sDr/z2Je1bvAcsCn8weiVX3D0FZdQNu+dcv2LTvrE9/U7jW2nUVIxZCESOi2bvtuFhvwvELwS/bUqUzY+rb+2CysVj/p2Hoo5J5/R69VTJ8MvsqpCXGYsZ7+/HfY751SYd0Ba5o8vQNl+D7okrM+/QwvnlqNKQiz1YvCiV1rRFXZCSFOww8d1Murs/tji2/leOzA2p8uP8s+nWT456h6bhzSE90U4S+nIQ3zlY3oGDzAfx+thZ3XdETf7n9Umef65DMJDy95RAWbDuC/x2vxCt3D/QqifMXVMOZ+AFuEheN5edYbA4UntFix8kq7DhZ5azVNSJLiYKxORiRlRzwfdY1WDHtnX2orDfjpbGpyO3R9mqDnkhNiMHHs0Zi5nv7MXtDIZZPHIi7hnhXq4wSfztkUhFeuvMyzHzvV7z542n8eVxOuENqxu5gUVFrwoSB4W/xCwQMRmQlY0RWMpbeNgBfH67AlsJyLPvPcfztuxO49pJuuCcvHdf179ZsuFq4sSyLbQfUWPx5ERgG+Ofkwbh9cM9mz0lNiMEHDw7D+7vL8Mq3x3HjP3bibxMH4rr+ni3d6Wzxh6EypyuVQtqlL+6eq2lwJvrdpzQwWOwQCxnk9VJi/k39IRIwWLOzBJPX7sXIrGTMGZcTsLU6Giw2PPD+fpRUGfDOzDyobBq/3zNJJsHGh0fgkQ9+w58/PoR6oxUzr+7j8esp8Xfgmku64c4reuLfP53CLZf3wCWpinCH5FRZb4LNwUZEV48rRYwYk4dlYvKwTJy6qMcnheX49Pdy/FBcCZVcgjsG98S9V2Ygp3t4j2Wd0YrnPzuKLw+dx7DeSqyYNKjd0tYCAYMHR/XBVf2SUfDRQTz4/m+YOiITC28e0OYQPFdVegsYAMq48Cb+FLkUxSHozogUJqsde0uqncm+pMoAgJvzcueQnhiT0w0j+yZDLm1KgVNH9MLGfWfx759O4941e3B1v2TMGZuDvN6+nwDMNjtmrS/EwXO1WHX/EIzOTkFxsf+JHwDkUhHenXklnvzwAJZ+eQx1RhuevL6fR9dEKfG7sejWAdhxsgrzPj2MTx+9KmzT7lsKZzlmT/XrJsf8m/rj6RtysONkFbb8Vo73d5fh7V9KMSgjEfcMTceEQWlIiG1/KFsw7C+twZzNB3Gh3oSnb8jBo9f08+jftX9qPD57/Gq89v0JvPVzKXafrsY/J12By9MT2n1Nlc6MhBhh2FcoU8klnXocP8uyKNEYsOMEl+j3llTDbHNAKhJgRFYypg7vhTGXpCBLJWs3McaIhfjTqD64b1gmNu47g9U7TmPi6j0Yna1CwdhsDO3l3QnAZnfgqQ8P4uc/NFg+cSBuvMz9IlHeihELser+IZj36RG8/sNJ1BotWHTLAAjcfJ4p8buhlEmwZMIAPPXRQazbXYYHR3n+dSqYwrkAi7dEQgGuz+2O63O7o1pvxrYDanxSWI7nPzuKF746hhsvS8W9eRkYmZXs9gPrD6vdgX/+8AdW/XQKGco4fProVRickejVe8SIhVh4ywBcc0k3zP34EO5ctQtzGsuItHXy0OjNSIwN//UhlVyKepMNZps9Iq9Xectic+BsTQP+qNThy1+rcPiLCpQ3TmjMSpHh/sZEP7yPEjFi7/7eWIkQD43Owv3De2HDXu4EcPe/uRPAnHE5GJLp/rqaw8Fi/tYj+LboAhbfOgD3+Fg63hMioQDLJw5EfKwI7+0qQ73Rhlfvvrzj1wQtmk7ktkFp2HZAjeXfncCh8loIGAYMAwgYBkKGgUDADWsUMEBdbS2S/yiCoPF3oYBxPsbfV0hFyO4uR//UeHSPl/o0XJWftRvJLf62JMuleGh0Fv40qg+OqOuw5bdyfH5Qjc8PnkfPxFiMze2GkX1VGJGlRGIAu0fKNAY8tfkgDp2rxb156Vgy4VLIpL5//K/up8K3BaOxcNtRLP/uBHacqMJr9w5yrkrG0+jNSIoJf6JtGssfnvWZfcGyLC7qzCipMqBEo0dplQElGgNKqvQ4pzU6yxbHihiMyumG2WP6YkxOSqt/A1/FSoR4OD8L94/IxPo9Z7BmZwnuWrUbY3JSUDA2G1e0cwJgWRYvfl2MTwrL8dT12SFpLAoEDBbfOgBJcRKs+O9J1JusmH1p+3mFEr8HGIbBS3dejoKPDuDQuVo4WMDBsnA42Kb7jT+tNhuYsgaX7Y2POZruu0qIFeOSVAX6pyqcP3O6KzqcyQdwXT3JMonbPuZIxTAMBqYnYmB6Ihbekovvj1Vi2+/l+Pi3cqzbcwYMA1yWloCr+iZjZN9kXNlb6VOiZlkWWwrLsfSLIoiFAqy6fwhu9mBdZk8kxknwxn1X4Lrfu2HJF0W4+Z8/4693XIo7Bvd0nsw1ejP6JYb/3yjFpWxDpCV+g9mGUo0Bp6v0KNUYUFJlQKmGu7kW/4sRC9A7WYZL0xJw68A0ZKXIkJUih6DuPAZeNiBo8cVJRJg1pi+mjuiF9XvPYM2O07hz1W5cc0kKCsbmtPrW+K/tp/DurlI8cHVvFASxzExLDMPgyeuzER8jwtIvj2H2pantPpcSv4d6JsZiy+yr3D6vuLgYubm5HT5Ha7DgRKUOJy7ocPyCDicu1GPr7+pmH/KeibHOkwF3QohHVorMOSqmXGsM/YVdiwGoUwP15UBduct9NQAWSOoDKPsAyqym+xL3Y5VjxELcNigNtw1Kg8XmwKHyWuw+VY3dpzV4b1cZ1uwsgUjAYHBGIrITWNwmqcYVmYluv8LXNliwcNtRfH2kAiOzkrFi0qCAz3JmGAZ3D03HsD5KzNl8EHM2H8L24ot46Y7LkRAnhkZnwZWp8oDu0xeBmr3Lsiz0ZhuMVjvMVgdMVjvMNu6nqfF3k63xMVvTNrPVDpPNwf1sfOxivRklGj0q65tiYhjus5+VIsfQXknISpGhj4pL8D3iY9rsCizWV/j1N3lKJhVh9pi+mDaiF9btKcPanSW4481duK5/NxSMzcbA9ES8+0spXv/hJCYOTceiWwaEdPIpb+bVfZCWGAuY2q80HB2J/5VMgBFwNzCN95l2trW1XQCIpEBMPCDlbwqX3xXNH4tp3CaNByRyQNDGhTmWBaxGwKxrvNUDZh3k5cWA+SBg0Tu3tbwlCSUY0WskRvTOB4ZdAYgkYFkW5VojTlzQ4URl0wlhx8kq2Bq/JoiFDPqmyJHTXYFj5+sxPCsww80AADYLoDvfmMzVQN05l/uNv5taThNnAHl3IKFxCOSxzwFjiwlv8u4uJ4Is7mTAnxTiWscvEQlwZW8lruytxFNjs2G02FF4RotdpzXYfboam4/U4sPDeyEVCZDXOwlX9VXhqr7JuDxNAZHdBFgbAJbFnkoh/rzlEDR6M+bf1B8Pj84K6oX5DGUcNs8aidU7TuP1/55E4Rkt/nr7ZTBa7RHSx9969i6fxLUGK2oaLNAaLKgxWHCitBZfnT2OGoOV29b4mLbBAm2D1eeVoWLEAsSIhZCKuJ9KmQSj+qVwLffG5N4rOc7zPnmWBYxaCCz13P9FobTt/6sBJpOK8Ng1/TB9ZG+s212Gt34uwW1v/IKRvRTYc0aH8Zem4pW7Lg/q9Sp3brg0FYWF7Sd+ho3wKZaFhYUYWrkZAAuwDu4fm3VwN+c2tPjd5XlgAYcdsJm5RGzik3Ed95N1V4uFcTkJxAFWU1NCZz0oUSAQu5xIGt/HqAUuNlYAFccBmSOA3qOBPvlAj8GAsOl8bLbZUVJlaPbt4MQFHc7XmfDM+Evw+LX9vDugViNQWQRUHAQqDsFYVohYSzWgr+SOlavYJCA+nUvsCelAfIufih6AqEU/vLEW0JYCNSVATWnj/cab7nzz58YkNP92kNSb225pAKwG7huG834DYDGgXnsRrN0Gc4MODoseIrsJcTAjjmnekm1gpagQpkKV0R8JPXMaTzqN+0pIBwTBS8aHy2tR8NFBlGi4IYRzR6XgiVuHBW1/njBZ7ei/6FtkKGMhk4hQ05jIrfbW//3jYEK6oAY5MbXIkmjRS1SDNFSjG6uB0nYRcfY6mKTJMMV0hyUuFVZZD9gVPcAq0sDE9wCTkA6xQoUYiRgxIiGkYgGkIoH3rV9LQ/NGSF3jN03Xb5y2Fqu/CSWAKIZr6Aml3E/+93Z/SgGBiMsRdjPXCGrzpxmwW1r9ZG1mMHbu86cTxEPWOw+CnkOAtCu4W3wa1yBt5EmvgN9YFoW//46hQ4e2+XB0JP52gvcby3LJpeUJwXm/vvl9i55L1M4k7pLMG++XnNcgq//Apm2idmZsGqqBM78ApT8DZb8AVcXcdokcyBwJ9BkN9B4FpA5qdiLgNVhsiBULO/7PZNYBF44AFYeablUnmk5YsUkwKLIg65nbmOAbkzyf7D3opvGK1QhoyxpPBCUuJ4USoPZs6xMpI+RiEMdxJ12JDA02BnEJqsZtMpiYGJxvEKCsHvhD68A5gwACOHBDDxNGJNVBVNu4P7vLiUEg5k4yfLeU60khMbP1ycwHDRYbXvq6GJv2n8U/b+mJ20YN9u8NWbZZ8vDFnzcfhLpGh15SPXqLatBTUI1UtgrJtiokWCshM1VAargAoVnb/IWMAFCkNX0+YpWAoQqoP8/ddBWt/+2EEq5hEJ/mcuvJ/VQ0/g60SOTlTd8u68pbf3sEAyhSmxoeCelAfBouVF5EanICl4htpsakbGr6vdnPtraZuMahSNJ4smj5U9p4QnH9KW06aTRuszMiCOrOgak4AFQeazomsm5NJ4G0K3DSIEfOkFF+/Vs6NdRw/3+qTwM1p4HqU433S1B4w2eU+EPF57O5vgooazwJlP0MaE5y26XxLieC0UDq5W23Vhtqmif4C4e5DwFPngr0GNT8lpCO4uPHg9/68ITdxrXuGEFTshdJWyU7d8f3fK0ROpOt+WQ7h4P7tsF/C6kpaf6NxOJShI0RAAkZ3EkhVskda4GIOwkJBI0/Rdx2Rtj4uLDFdoHzdRYHUHvxPLolxTcmGlPzhGN1/d3YPCFZXX53JlaX7syW3Z4tuzxbPZfhPictk3RMAvc3N36bu2iRolu/IU3JVdGjzcZH0/G1N54I1E0ng2b3G292c/vvAQDShKZvl85vlhlNJxxFWpsn5ZC0oL3Ff7M+f6DpVnW8qYdBkeZyMhjM/ZSp2n4vs84lsZc0T/CuJ0f+s5vcF1D2RWHqlHZzZ3T08XcF8hTgsru4GwDoLjSdBEp/Bv74jtsekwD0upq7WY2NXTaHgbqzTe+VkAn0GAgMnNyY5AdyLaVIJhQBSb38fpu0tkasCARNyaRPfvPHWJZLWq1OCiVcq9Nh5xKlw+5y38adTJzbbdz9NroNJQC6Of9Gl+4FcYxLV0Ms9zNO1fhYrEtXRONNIALXlcm26NZ0/b3x1uZjjb/HKRuPRVOiR0zzujHVxcXo5k0iFQi5z5ciFejZTiONZbmTjvOEoOZORAkZjQm+J/fZ7izEsUB6HnfjWQzAhSO4cOBbpNrPcyeDE9/A2cWakMGdBFL6c///+WSvb1GILb4n9w11wO3OJI/kftz/H9cehsLCdsOjxB+pFKnA5RO5G8D9Zyn7BSjdyf088Q23PbkfkHElMOwhLsmnDmzzoilpB8MA8m7cLXOEf+/Fss1PFI0/j58qRf9LB4XkwmPEYhhAlszdegwMdzThIZEBmSOgNSQglT+xmuq5b+eu3wyKv+S6h5L7AdnjGhM7n9z7cN2efqLEHy3i04CB93I3gGsRiONatdZIGDFMY5dI8/9WrCimayd90r6YeO5aXm+XPn+7reOutQCgxB+tIr3rhhDimyAnfYAWYiGEkC4n7C1+h8OBpUuX4sSJE5BIJHjxxRfRq5f/F/kIIYS0Lewt/h9++AEWiwWbN2/G3Llz8corr4Q7JEII6dTCnvgLCwsxevRoAMDgwYNx9OjRMEdECCGdW9i7evR6PeTypiJWQqEQNpsNIlFTaMXFxeEIzScmk4niDaJoixeIvpgp3uCKhHjDnvjlcjkMBoPzd4fD0SzpA4i8WXkdiMhZhB2geIMv2mKmeIMrVPEWdjCBK+xdPUOGDMHOnTsBAAcPHkROTmQtak4IIZ1N2Gv18KN6Tp48CZZl8fLLL6Nv377Oxzs6axFCCGlf1BZpI4QQElhh7+ohhBASWpT4CSGkiwn7qB6euxm8H3/8MT766COIRCI8+uijuPbaa8MYLWC1WrFgwQKo1WpYLBY8+uijuP76652Pv/fee/jkk0+gVHKVMv/yl78gKysrXOECAO644w4oFFyd+vT0dCxbtsz5WKQd361bt2Lbtm0AALPZjOLiYuzatQvx8VxRuhdffBG///47ZDJusZhVq1Y5/7ZQO3ToEP7+979j/fr1OHPmDObPnw+GYZCdnY0lS5ZA4FKgzWQy4ZlnnkF1dTVkMhleffVV52ckHPEWFxfjhRdegFAohEQiwauvvgqVqnld+I4+N6GOt6ioCLNnz0bv3r0BAFOmTMHNN9/sfG6kHd85c+ZAo9EAANRqNQYNGoTXX3/d+VyWZZGfn+/8ewYPHoy5c+cGP0g2Qnz33XfsvHnzWJZl2QMHDrCzZ892Pnbx4kX21ltvZc1mM1tfX++8H06ffPIJ++KLL7Isy7I1NTXsmDFjmj0+d+5c9siRI2GIrG0mk4m9/fbb23wsEo+vq6VLl7IfffRRs22TJ09mq6urwxRRk7Vr17K33nore88997Asy7KzZs1i9+7dy7Isyy5atIj9/vvvmz3/3XffZf/1r3+xLMuyX331FfvCCy+ENd7777+fPXbsGMuyLPvhhx+yL7/8crPnd/S5CYWW8X788cfsO++80+7zI+348mpra9nbbruNraysbLa9rKyMnTVrVihDZFmWZSOmq6ejGbyHDx/GFVdcAYlEAoVCgczMTBw/fjxcoQIAbrzxRjz11FPO34XC5qtiFRUVYe3atZgyZQrWrFkT6vBaOX78OIxGIx588EFMnz4dBw8edD4WiceXd+TIEZw6dQqTJk1ybnM4HDhz5gwWL16MyZMn45NPPglbfJmZmVi5cqXz96KiIgwbxq2vm5+fj927dzd7vuvnPD8/H3v27AldsGgd74oVK5xjyu12O6TS5kuFdvS5CUe8R48exU8//YT7778fCxYsgF6vb/b8SDu+vJUrV2Lq1Kno1q1bs+1FRUWorKzEtGnT8PDDD6OkpCQkcUZM4m9vBi//mOvXeJlM1uofPNRkMhnkcjn0ej2efPJJFBQUNHv8lltuwdKlS7Fu3ToUFhbixx9/DFOknJiYGPzpT3/CO++8g7/85S94+umnI/r48tasWYPHH3+82baGhgZMnToVy5cvx9tvv41NmzaF7UQ1fvz4ZhMOWZZ1roMsk8mg0+maPd/1WLf1eLC1jJdPRL///js2bNiAmTNnNnt+R5+bcMQ7cOBAPPvss9i4cSMyMjLw5ptvNnt+pB1fAKiursaePXtw1113tXp+SkoKHnnkEaxfvx6zZs3CM888E5I4IybxdzSDt+VjBoMhbP25rioqKjB9De5vggAABXpJREFU+nTcfvvtmDBhgnM7y7KYMWMGlEolJBIJxowZg2PHjoUxUqBPnz647bbbwDAM+vTpg8TERFRVVQGI3ONbX1+PkpISjBjRfGWs2NhYTJ8+HbGxsZDL5RgxYkTEfENx7c83GAzOaxI812Pd1uPh8M0332DJkiVYu3Ztq/7wjj434TBu3Dhcdtllzvst/19F4vH99ttvceutt7bqFQCAyy67zHltMC8vD5WVlWBDMMI+YhJ/RzN4Bw4ciMLCQpjNZuh0Opw+fTrsM3w1Gg0efPBBPPPMM5g4cWKzx/R6PW699VYYDAawLIt9+/Y5P6zh8sknnzgrn1ZWVkKv1yMlJQVAZB5fAPj1119x1VVXtdpeVlaG++67D3a7HVarFb///jsuvfTSMETY2oABA7Bv3z4AwM6dO5GXl9fs8SFDhmDHjh3Ox9ubYBMqn3/+OTZs2ID169cjIyOj1eMdfW7C4U9/+hMOHz4MANizZ0+rf/dIO74AF2d+fn6bj73xxhtYt24dAK5bLS0tzfmNMZgiZlTPuHHjsGvXLkyePNk5g/e9995DZmYmrr/+ekybNg333XcfWJbFnDlzWvVFhtrq1atRX1+PVatWYdWqVQCAe+65B0ajEZMmTcKcOXMwffp0SCQSjBw5EmPGjAlrvBMnTsRzzz2HKVOmgGEYvPzyy1i/fn3EHl8AKC0tRXp6uvN318/DhAkTcO+990IsFuP2229HdnZ2GCNtMm/ePCxatAgrVqxAVlYWxo8fDwB48MEHsXr1akyZMgXz5s3DlClTIBaL8dprr4UtVrvdjpdeegk9evTAE088AQC48sor8eSTT+LZZ59FQUFBm5+bll0ZobR06VK88MILEIvFUKlUeOGFFwBE5vHllZaWtjqp8vE+8sgjeOaZZ7Bjxw4IhcKQjZiimbuEENLFRExXDyGEkNCgxE8IIV0MJX5CCOliKPETQkgXQ4mfEEK6mIgZzklIOOzbtw8FBQXo168fAK4g3IQJE3D27Fk88MADSEtL8+r9LBYLFi5ciFdffbXZZC7ehx9+iN69e2PkyJEBiZ8QX1DiJ13eiBEjnBUTLRYLbrzxRnz22Wc+zfp8//33cdNNN7WZ9AFurscDDzyAYcOGtTmTk5BQoMRPiAu9Xg+BQICZM2di+fLl+Oabb1BeXo7q6mqcP38ezz33HEaPHo0ff/wR//rXvyCXy5GQkIBLLrkE//d//4cvvvjCWU769ddfx969e+FwOHDLLbdg5syZEIlEuPTSS/HTTz81K+NNSChR4idd3t69ezFt2jQwDAOxWIxFixbh7bffdj4ukUjw9ttvY9euXXj33Xdx1VVX4cUXX8TmzZuhUqmc9dPLysogl8shFosBAJ999hk2bNiA7t27Y+vWrc73u+SSS7B//35K/CRsKPGTLs+1q4fnmvj5ssWpqamwWCyoqamBXC53LliSl5cHjUYDrVbbbBGTFStWYMWKFdBoNM5SwQBXkXHv3r3B/JMI6RCN6iHEjZZFs5KTk2EwGFBTUwOAW3GJ315fXw+Au1bw7bffYsWKFVi3bh22bdsGtVoNgKs6GupVoQhxRS1+QrwkEAiwaNEiPPzww1AoFHA4HOjVqxd69eqFmpoa2Gw2SCQSJCQk4Pbbb0dCQgKuvvpq5wihQ4cO4eqrrw7zX0G6MirSRogP1qxZgwceeAASiQRPP/00Ro0ahTvuuANr1qxBVlYWxo0b1+brbDYbHnjgAbz//vs0qoeEDXX1EOIDmUyGe++911lGnF/we8aMGfj222/hcDjafN3mzZsxa9YsSvokrKjFTwghXQy1+AkhpIuhxE8IIV0MJX5CCOliKPETQkgXQ4mfEEK6GEr8hBDSxfw/JaH/7DBGvG8AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_fig.plot()\n",
    "plt.legend(loc='best')\n",
    "\n",
    "plt.xlabel('Ping(s)')\n",
    "plt.ylabel('Runtime (secs)')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><h2 style=\"font-size:14px\"style=\"padding:1px\"><center>Continued</center></h2>\n",
    "<p> \n",
    "The more user stories the better the 'final-release' product is. The more options to visualize all that is possible with a certain data the more informative, significant and trustworthy the science created from this product will be. And able to withstand critic, challenges and time due to its intrinsic design, that always tries to prove all its components wrong. \n",
    "Another important future aspect that needs the data to be handled with excessive care, review, standards, enrichment, principal components-per-sample definitions and so many others due to the possible future machine learning structures formed by this ocean of information that can then be used as features and layers for an AI to look for what is there. There might be a never seen correlation between two networks that have no equal nodes (disease, tissue, etc) but that when compared in its distances overlaps parts of one network onto of the others.<p>\n",
    "<p>\n",
    "A possible idea is to look near the overlaps location. Maybe there is a layer in the first network that defines color, where is that layer anchored and what is the direct corresponding anchor on the overlay? Does the overlay node has anything to do with color? <p>\n",
    "Everything evolved from a branch you can follow the evolution of a structure from the simplest life almost forms to complex animal life such as our own. This means the branches, its directions and the point from where they came from were once the same. A turtle lives a really long time, give all the information for that turtle tissues, protective carapaca and its protein, gene and its orthologs, expression, response to drugs and even to temperature all this as the beginning of a massive amount of accumulating layered network. Which each layer main node being structure as a dictionary with growing values assign to one key. All these keys are main values of other dictionaries.\n",
    "<p><p>\n",
    "It will be a ball-like universe on nodes, connections, linking path up/down and transversally. Now please AI use all the computation we can afford what does this turtle have that I don't. Give me all results and their 10-step downstream and upstream information. Isolated and create this as a new cosmos-like cluster of aging. Let's separate in modules and test.\n",
    "<p>\n",
    "*Ps.: For those who feel fatigue, D-Ribose is a good supplement, takes a different route and creates more energy for those lacking enough ATP/Energy production due to mitochondrial dysfunction, unbalanced diet and/or aging. <p>\n",
    "*Curiosity instigating blurb:<br>\n",
    "<p> ATP molecule is able to store and transport chemical energy within cells and is essential for synthesis of nucleus acids such as DNA and RNA. Ribose is a naturally occurring five-carbon sugar produced in the body through the PPP, a metabolic pathway parallel to glycolysis that generates nicotinamide adenine dinucleotide phosphate (NADPH), pentoses, and ribose 5-phosphate. The PPP is a slow process that requires an enzyme called glucose-6-phosphate dehydrogenase (G-6-PDH) that is often in short supply within the cells. This enzyme can have limited expression in the myocardial cells with cardiac disease leading to significant delay in the production of ribose.<p>\n",
    "In certain pathologic conditions such as heart failure, cellular energy deficiency exists in myocardial mitochondria. The reduction in ATP production is directly correlated with the decreased supply of D-ribose in the mitochondria. This may be related to the limited expression of the G-6-PDH enzyme in myocardium that can significantly interrupt the production of ribose. Several studies have shown that augmenting D-ribose following myocardial ischemia improves mitochondrial function by increasing myocardial ATP production.\n",
    "<p>\n",
    "Administering supplemental D-ribose circumvents the enzymatic step to assist with replenishing ATP levels in cells. In purine metabolism, D-glucose is transformed to both D-ribose-6-phosphate and to D-ribose-5-phosphate and then altered to 5-phospho-D-ribose 1-pyrophsophate (PRPP) for the synthesis of purine and pyrimidine. In other words, supplemental D-ribose bypasses the rate-controlling PPP (slower pathway) and provides an alternate source of PRPP for ATP production. <p>\n",
    "<p>\n",
    "D-ribose has been used both orally and intravenously in patients for many different pathologic conditions such as chronic fatigue syndrome, fibromyalgia, and myocardial dysfunction. It is often used to improve athletic performance and reduce symptoms of cramping, pain, and stiffness following exercise. Under different pathologic conditions, ATP, ADP, and adenosine monophosphate are degraded and not available for energy production. Supplemental D-ribose has been shown to enhance recovery of ATP levels and reduce cellular injury in humans and animals.\n",
    "<p>\n",
    "IN SUM: More energy, less felt effects of multiple diseases, less oxidative stress and a general feel better metabolism. Without a productive and efficiency energy metabolism, the body and its metabolism will start feeling and eventually will evolve to a system-wide adverse condition. <p>\n",
    "<p>\n",
    "Everything that can be improved in the metabolism should be since the accumulating metabolic sub-products are a main driver of aging, disease and cellular disfunctions.<p>\n",
    "*Another example when the permanent teeth take over, a signal from its base tells the entire body \" It's time to get older, lets change the production to a bit older.\" It’s a maybe absurd idea but why not reduce that signal. Tooth I feel you might need to be stimulated into forgetting a permanent role, after all by 75 years old a new teeth-set would be welcomed.\n",
    "<p> \n",
    "*All in all, I would be willing to take a meal a day as a shake with 1000 different substances, calculated and mixed to help my body to function optimally. The 10 or so  pills I will happily take with added-extra-magnesium water, all in all being able to guarantee better brain function, feeding the brain easier with all co-factors, vitamins and even neurotransmitters precursors to supplement the non-optimum concentration of neurotransmitters as people age, especially after reaching 30 years-old. This is one of the reasons some people do not respond to antidepressants and other mental health medications that need to use/force the brain machinery to produce more neurotransmitters, if the machine is available but the input/the dirt to the neurotransmitters-bricks is not available and the necessary quantity the medication either does not work, needs a higher dose or stops working overtime. This is the main reason for medication-resistant depression.\n",
    "<p>\n",
    "The secret to longer healthy life is just a gigantic database away. Given enough computation, data curation, natural language processing and multiple in parallel instances running ML algorithms to connect information and then AI in parallel as highly-powered computation, informed-best-possible-decision-able genius scientists that never need to stop, this is not an out of reach goal, and a decade would be enough.\n",
    "<p>\n",
    "The road-map of the necessary work is ready. The work has to start and amazon needs to be kind and help fund the computation. The outline project is already written. Heck at this point I admit I'm getting carried-away but I happen to have read all of geo aging paper until 2017 and created aging collections, I have studied so much that it is only fair to say it’s possible I bet my life on it. Think of the implications for space-travel.\n",
    "<p>\n",
    "* Also given a target and knowing the necessary altered target-specific response coupled with the pathway up and down information any place can be hit by a design small molecule to hit the target, these changes are only to reduce, increase, block and delay the simplest target interactions, the signaling changes can be further studied in the future but specific target can be modulated in a simple pharmacology designed molecules with chosen degradation and/or re-uptake. \n",
    "<p>Every part of the metabolism can be targeted and hit-specifically the downstream effects and redundancy are the main issues which are to be given to the AI to figure out and give out a list of targets and their necessary modulation to achieve a final overall response.\n",
    "<p>\n",
    "<h5> Observation about this issue:</h5> Guess the above is my user story, get informative, high quality data, find meaning, enrich, inform, add more information and give it to ML and AI with directed research focus on aging and cognitive function. This is my research dream-goal and it will happen even if I have to find more ways to get funding, write the project (it’s a project with 15 sub-projects inside and identified government databases for background data (over 100 databases drug information, toxins and plant detailed 'medicinal' uses that can and are increasingly being characterized and the beneficial compounds then isolated and further developed before final-use-result (that is already known as part of the traditional medicine as effective being given higher efficiency, efficacy and less adverse effects due to the break-down of the traditional medicine keeping only the beneficial compounds, the rest is taken out of the developed drug/medication which increases its safety.<p>\n",
    "    <p>\n",
    "* My own anecdote I know personally and seen the effects over time of a mix of plants being a cure for vitiligo. As long as the concoction is taken it stops and reverts. I have seen the plant doctor and the patient with vitiligo and its reduction over time in two years it was gone and it will stay that way as long as he keeps drinking it, this was over 10 years ago. The reasoning behind the manipulated plant-mix medications is quite absurd and somewhat based in witch-doctor culture, tales and books. Yet the outcome of it, the final result is proven and has been tested on people for decades. I knew three plant-doctors in my childhood, two have passed away and the last one has kindly agreed to write his formulas down. This took me asking since childhood and thus the man has asked for some money so he can retire as long as it’s not forcibly taken, he and a couple of Indians according to him will pass it over to someone who is interested in keeping the knowledge alive. The ones I've seen over time, probed into the 'patients' complains and evolution and the vitiligo pictures I myself took is why I can say with confidence there are at least 5 new drugs with no other known treatment in the man’s 'office'. <p> Full disclose: I was taken there by my grandmother at about 5 year old and took an interest in the mixes and the vehicles the mixes are added to (usually wine, random cheap alcohol and even belladonna paste) which is why I know the details of his formulas, preparations and the source for his supply of Amazonian (mostly) plants. History tells an Indian taught him. <p>\n",
    "    <p>\n",
    "With all that is known in science and what can be made into scientific-knowledge looking for the pharmacology effects of each sequenced plant compounds, all in order and structured, with more computation, automation of pipelines, training datasets multiple times until the algorithms become highly trustworthy for highly specific stratifications(data-type-assay-tissue-age-otherCharacteristics without a limit of variables added to this stratified highly specific segments that will then be defined as gold standards and used as reference for comparisons against newer generated data/analysis that can be given different significant values per characteristics-segment-length. Lower number of characteristics will created more correlations but higher numbers will result in higher significance, by looking at the multiple segment length results, cross sections can be drawn and a limit for significance determined for each global category (or the main full network regions).<p>\n",
    "    <p>\n",
    "While a lot of data will be necessary with detailed information, globally the necessary data can be easily collected from hospitals, volunteers, collaborators and the researchers that which to add their own data, some researchers might be asked to make their data public and identified for further data enrichment which for people that can be tracked will be continuously aggregating information.<p>\n",
    "  <p>\n",
    "At face value the entire idea of aging research is dismissed as too difficult and too laborious. What I propose is the alternative view that accepts the amount of work as a necessity and having taken steps to start the required work is now personally seen as not only possible but also containing and easy to map route, it will a lot of work along the way but the route is ready and highlighted in the map.<p>\n",
    "    <p>\n",
    "Seeing what exist available already a 10 years project to map the route of aging, defining the rest-stops (targets) along the way with fine granularity of detail seem possible that in 5 years ML algorithms will be finishing their data structuring and node connections, given 3 years for the general cosmos-cluster network assembly then at the 8 year mark of the project official beginning an AI will be running as a main higher hierarchy decision that will control the limits and guide pre-sets enforcing conditions defined as vital such as those conditions and limits to which human survival is subjected to and the window(ranges) acceptable for each category and then the limit per category_item thus constraining the search to those that are viable for human/mammals life without serious adverse effects (such as liver damage, accumulating toxic substances, non-acceptable targets (brain, brain-barrier, placenta-barrier) these need to be taken to a round table of human specialists in the target location medicine, researchers, biologists, pharmacologists and biochemistry which will each give a report as pros and cons according to their own science and knowledge, guaranteeing a multi-sided view that will then be used to create a detailed report that should be submitted to and ethics committee along with the round-table target reach-hit-modulating strategy for ethics review, once approved the authorization to continue target investigation given the agreed as per round-table and ethics committee approved limits will be given to the main AI, documented and registered with proper ethical and legal protocols. <p>\n",
    "    <p>\n",
    "    *Once an decision is approved by the round-table (containing name, specialty, specific-report-ID created by the individual that responds as responsible for that specific report-ID, each individual responsibility is limited to the information documented on the written report, this person can request a second, third up to five opinions of specialist who shall answer the same questions and provide each one report the majority on each field is them defined as the final report.) The specialists on each area should be chosen previously in sufficient number, with no upper limit just a minimal limit defined as over 12 specialists per area. These steps; creating of the round-table and of the ethics committee are required to fulfill the applicable laws, guidelines and The WHO ethical requirements and good research practices. <p>\n",
    "<p>\n",
    "At the 10 year mark the target list should contain multiple targets and their metabolic result. As the target list grows and the result achieve moves closer to the desired outcome new treatments possibilities for multiple diseases will be 'found' as possible from those a secondary research focus area will be explored for disease-specific treatment possibilities.\n",
    "<p> <p>\n",
    "At the 15 year mark the supplement shake and pills should be going to phase III trials. Phase I and II trials are accelerated by the limits set as human survival window that is enforced at the highest hierarchical level. Necessary Phase I and II test are part of the predictions and also help guide the AI choices (teratogenic compounds for instance are banned from use, maximum dose per substance type-effect-class are also predicted (saturation is calculated using the substance chemical characteristics) the pharmacopeia which contains descriptions and reference values/limits for all tests a drug needs to be subjected too to be safe, bioavailability (AUC), determine how long the good for consumption period, packing type depends on medication compounds sensitivity mostly to light and is used to define package color and material. The dilution time of a pill and its resistance to applied force until break, spectrophotometry and gas chromatography values are also predicted based on chemical characteristics all these are part of the background knowledge and basing information that will be available to inform and guide decisions.\n",
    "<p> <p>\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
